<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Preparing raw Illumina data in different formats for use with QIIME &mdash; Homepage</title>
    
    <link rel="stylesheet" href="../_static/default.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '1.9.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <link rel="shortcut icon" href="../_static/favicon.ico"/>
    <link rel="top" title="Homepage" href="../index.html" />
    <link rel="up" title="QIIME Tutorials" href="index.html" />
    <link rel="next" title="Analysis of 18S data" href="processing_18S_data.html" />
    <link rel="prev" title="Fungal ITS analysis tutorial (an IPython Notebook): open-reference OTU picking" href="fungal_its_analysis.html" />
<meta http-equiv="Content-Style-Type" content="text/css" />
<script type="text/javascript" src="http://www.google.com/jsapi?key=ABQIAAAAbW_pA971hrPgosv-Msv7hRRE2viNBUPuU405tK6p2cguOFmlFBQSwZMG6_q_v6Z42nkdo9ejT1aHmA"></script>
<script type="text/javascript" src="../_static/google_feed.js"></script>

  </head>
  <body>

<div style="background-color: white; text-align: left; padding: 0px">
<a href="../index.html"><img src="../_static/wordpressheader.png" border="0" alt="sampledoc"/></a>
</div>
<div class="news" style="background-color:#ccc;"><table style="font-size:12pt;padding: 5px;"><tr id="feed"><td><b>News and Announcements &raquo;</b></td>
</tr></table></div>

    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="processing_18S_data.html" title="Analysis of 18S data"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="fungal_its_analysis.html" title="Fungal ITS analysis tutorial (an IPython Notebook): open-reference OTU picking"
             accesskey="P">previous</a> |</li>
        <li><a href="../index.html">Home</a> &raquo;</li>
          <li><a href="index.html" accesskey="U">QIIME Tutorials</a> &raquo;</li> 
      </ul>
    </div>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="preparing-raw-illumina-data-in-different-formats-for-use-with-qiime">
<span id="processing-illumina-data"></span><h1>Preparing raw Illumina data in different formats for use with QIIME<a class="headerlink" href="#preparing-raw-illumina-data-in-different-formats-for-use-with-qiime" title="Permalink to this headline">¶</a></h1>
<p>The <a class="reference external" href="./illumina_overview_tutorial.html">Illumina Overview Tutorial</a> describes how to work with raw Illumina sequence data with QIIME. That tutorial covers the case where you are starting with three specific input files: your <a class="reference external" href="../documentation/file_formats.html#metadata-mapping-files">sample metadata mapping file</a> which contains the per-sample barcode sequences, a <tt class="docutils literal"><span class="pre">fastq</span></tt> file containing your amplicon sequence reads, and a corresponding <tt class="docutils literal"><span class="pre">fastq</span></tt> file containing the barcode reads for each amplicon sequence. If you don&#8217;t have your data in this format, this document will help you get your data into this format for use with QIIME.</p>
<p>Demultiplexing and quality filtering of Illumina data is performed using <a class="reference external" href="../scripts/split_libraries_fastq.html">split_libraries_fastq.py</a>. The other scripts described in this document will help you to get your data into <a class="reference external" href="../scripts/split_libraries_fastq.html">split_libraries_fastq.py</a>.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">The preferred format in QIIME for Illumina data is <tt class="docutils literal"><span class="pre">fastq</span></tt>. A detailed understanding of the <tt class="docutils literal"><span class="pre">fastq</span></tt> format is not necessary to use QIIME, but you can find details <a class="reference external" href="http://scikit-bio.org/docs/latest/generated/skbio.io.fastq.html">here</a> if you&#8217;re interested. If your data is not in <tt class="docutils literal"><span class="pre">fastq</span></tt> format, this document can help you convert from some other less common formats to <tt class="docutils literal"><span class="pre">fastq</span></tt>.</p>
</div>
<div class="section" id="quality-filtering-of-illumina-data-with-qiime">
<span id="illumina-quality-filtering"></span><h2>Quality filtering of Illumina data with QIIME<a class="headerlink" href="#quality-filtering-of-illumina-data-with-qiime" title="Permalink to this headline">¶</a></h2>
<p>Quality filtering of Illumina data in QIIME is performed as part of <a class="reference external" href="../scripts/split_libraries_fastq.html">split_libraries_fastq.py</a>. Details on the quality filtering strategy in QIIME, including motivations for the default parameter settings, are described in <a class="reference external" href="http://www.ncbi.nlm.nih.gov/pmc/articles/PMC3531572/">Bokulich et al. (2013)</a>.</p>
<p>If you only wish to perform quality filtering of your fastq sequences (and not demultiplexing), you can use <a class="reference external" href="../scripts/split_libraries_fastq.html">split_libraries_fastq.py</a> with the <tt class="docutils literal"><span class="pre">--barcode_type</span> <span class="pre">'not-barcoded'</span></tt> option.</p>
</div>
<div class="section" id="demultiplexing-and-quality-filtering-fastq-illumina-data-with-qiime">
<h2>Demultiplexing and quality filtering fastq Illumina data with QIIME<a class="headerlink" href="#demultiplexing-and-quality-filtering-fastq-illumina-data-with-qiime" title="Permalink to this headline">¶</a></h2>
<p>The standard approach for demultiplexing a single lane of Illumina fastq data with QIIME is illustrated in the <a class="reference external" href="./illumina_overview_tutorial.html">Illumina Overview Tutorial</a>. Special cases are discussed here.</p>
<div class="section" id="starting-with-gzipped-fastq-files">
<h3>Starting with gzipped fastq files<a class="headerlink" href="#starting-with-gzipped-fastq-files" title="Permalink to this headline">¶</a></h3>
<p><a class="reference external" href="../scripts/split_libraries_fastq.html">split_libraries_fastq.py</a> can work with either gzip-compressed (e.g., <tt class="docutils literal"><span class="pre">.fastq.gz</span></tt>) or uncompressed (e.g. <tt class="docutils literal"><span class="pre">.fastq</span></tt>) <tt class="docutils literal"><span class="pre">fastq</span></tt> files. You can pass compressed files to <a class="reference external" href="../scripts/split_libraries_fastq.html">split_libraries_fastq.py</a> in the same way that you pass uncompressed files.</p>
</div>
<div class="section" id="processing-multiple-lanes-of-illumina-fastq-data-with-qiime">
<h3>Processing multiple lanes of Illumina fastq data with QIIME<a class="headerlink" href="#processing-multiple-lanes-of-illumina-fastq-data-with-qiime" title="Permalink to this headline">¶</a></h3>
<p>If you have multiple lanes of Illumina data that you want to demultiplex together, you&#8217;ll use a similar command as for a single lane (illustrated in the <a class="reference external" href="./illumina_overview_tutorial.html">Illumina Overview Tutorial</a>), but will specify per-lane amplicon read, barcode read, and mapping files. To do that, call the same command with comma-separated filepaths. Note that the order of the files must correspond for each of these parameters. For example, to demultiplex data from lanes 6, 7, and 8, your command would be:</p>
<div class="highlight-python"><div class="highlight"><pre>split_libraries_fastq.py -i s_6_sequence.fastq,s_7_sequence.fastq,s_8_sequence.fastq -o sl_out/ -b s_6_barcode.fastq,s_7_barcode.fastq,s_8_barcode.fastq -m s_6_map.txt,s_7_map.txt,s_8_map.txt
</pre></div>
</div>
<p>When the command completes you&#8217;ll have the same output as when running on a single lane, but note that the data in the log and histogram files are broken down by lane. All of the sequence data will be in <tt class="docutils literal"><span class="pre">seqs.fna</span></tt>.</p>
<p>You may wish to review <tt class="docutils literal"><span class="pre">split_libraries_log.txt</span></tt> and adjust quality filtering parameters and rerun. When you&#8217;re satisfied, you&#8217;re read to move on to OTU picking, which is discussed in the <a class="reference external" href="./illumina_overview_tutorial.html">Illumina Overview Tutorial</a> and in <a class="reference external" href="otu_picking.html">OTU picking strategies in QIIME</a>.</p>
</div>
</div>
<div class="section" id="demultiplexing-and-quality-filtering-non-fastq-illumina-data-with-qiime">
<span id="other-file-formats"></span><h2>Demultiplexing and quality filtering non-fastq Illumina data with QIIME<a class="headerlink" href="#demultiplexing-and-quality-filtering-non-fastq-illumina-data-with-qiime" title="Permalink to this headline">¶</a></h2>
<div class="section" id="processing-illumina-qseq-files">
<h3>Processing Illumina qseq files<a class="headerlink" href="#processing-illumina-qseq-files" title="Permalink to this headline">¶</a></h3>
<p>You can convert qseq files to fastq files using the <a class="reference external" href="../scripts/process_qseq.html">process_qseq.py</a> script. The resulting fastq files can then be processed with <a class="reference external" href="../scripts/split_libraries_fastq.html">split_libraries_fastq.py</a>.</p>
<p>Example qseq file for amplicon read:</p>
<div class="highlight-python"><div class="highlight"><pre>M10     68      1       1       28680   29475   0       1       AACGAAAGGCAGTTTTGGAAGTAGGCGAATTAGGGTAACGCATATAGGATGCTAATACAACGTGAATGAAGTACTGCATCTATGTCACCAGCTTATTACAGCAGCTTGTCATACATGGCCGTACAGGAAACACACATCATAGCATCACACGA        BBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB        0
M10     68      1       1       19607   29475   0       1       GACATAAGGGTGGTTAGTATACCGGCAAGGACGGGGTTACTAGTGACGTCCTTCCCCGTATGCCGGGCAATAATGTTTATGTTGGTTTCATGGTTTGGTCTAACTTTACCGCTACTAAATGCTGCGGATTGGTTTCGCTGAATCAGATTATT        Z__c\JQ`cc[[_[bfff[[`Qbdge_YYOOHO^cF[FUb_VHMHV`T`dBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB        1
M10     68      1       1       22962   29475   0       1       TAATCGAGCTCAACGCCCTGCATACGAAAAGACAGAATCTCTTGCAAGATGTTGGTGCGGTTAGCCAGCTGCTTATGGAAGCCAAGCATTGGGGATTGAGAAAGAGTAGAAATGCCACAAGCCTCAATAGCAGGTTTAAGAGCCTCGATACG        JJY````JO[`bab`b`bbaaaaa`\`a`OVT``]]`aa^aI\HMMMWWHHNNNGLL\`________\Z^]]^^^^^^GX]\QTXXZ[YZ^^XZ[Z^\Z^GW\^^\\^^^VZ\Y^^^^\\\\[^[\\\^VWYWWXWWZYZW^[X^\\Z^[TQ        0
</pre></div>
</div>
<p>Example qseq file for barcode read:</p>
<div class="highlight-python"><div class="highlight"><pre>M10     68      1       1       28680   29475   0       2       ACTCACGGTATTA   \_J\Sa^Y[ZYK`   0
M10     68      1       1       19607   29475   0       2       AGACTGAGTACTA   PP\JJ\JQ`\RK^   1
M10     68      1       1       22962   29475   0       2       AGACGTGCAATTA   ^_aecceeeQ`[b   0
</pre></div>
</div>
<p>You&#8217;ll need to know which of your reads files correspond to your barcodes and which correspond to your amplicons. To determine this you can look at the first line of representative files with the <tt class="docutils literal"><span class="pre">head</span></tt> command. For example:</p>
<div class="highlight-python"><div class="highlight"><pre>head -n 1 s_1_1_0001_qseq.txt
</pre></div>
</div>
<p>Will give you the first line s_1_1_0001_qseq.txt. If the length of the sequence (the 9th field) corresponds to the length of your barcode, then this is your barcode read file. If not, check a qseq file corresponding to another read number (e.g., s_1_2_0001_qseq.txt). Note that due to technical artifacts you may sometimes have a single extra base here, so for a length 12 barcode your sequence may be length 13.</p>
<p>You&#8217;ll typically start here with a directory containing many qseq files. The process_qseq.py script therefore works on a directory, rather than a set of input files. In my example, the read 1 files correspond to my sequence reads and the read 2 files correspond to my barcode reads. To generate a single fastq file for the sequence reads from the qseq files, you can run the command:</p>
<div class="highlight-python"><div class="highlight"><pre>process_qseq.py -i ./ -o ./fastq/ -r 1
</pre></div>
</div>
<p>This specifies that the qseq files are in the current directory (<tt class="docutils literal"><span class="pre">-i</span></tt>), and the fastq should be written to <tt class="docutils literal"><span class="pre">./fastq/</span></tt>. The <tt class="docutils literal"><span class="pre">-r</span> <span class="pre">1</span></tt> specifies that I want to process the read one files (i.e., my amplicon reads).</p>
<p>To generate the barcode read fastq file you can run the following command:</p>
<div class="highlight-python"><div class="highlight"><pre>process_qseq.py -i ./ -o ./fastq/ -r 2 -b 12
</pre></div>
</div>
<p>This again specifies that the qseq files are in the current directory (<tt class="docutils literal"><span class="pre">-i</span></tt>), and the fastq should be written to <tt class="docutils literal"><span class="pre">./fastq/</span></tt>. The <tt class="docutils literal"><span class="pre">-r</span> <span class="pre">2</span></tt> specifies that I want to process the read two files (i.e., my barcode reads), and the <tt class="docutils literal"><span class="pre">-b</span> <span class="pre">12</span></tt> specifies that I only want to extract the first twelve bases of these reads.</p>
<p>Once these steps are complete you&#8217;ll have fastq files that can be passed to split_libraries_fastq.py.</p>
</div>
<div class="section" id="processing-iseq-files-with-qiime">
<h3>Processing iseq files with QIIME<a class="headerlink" href="#processing-iseq-files-with-qiime" title="Permalink to this headline">¶</a></h3>
<p>You can convert iseq files to fastq files using the <a class="reference external" href="../scripts/process_iseq.html">process_iseq.py</a> script. The resulting fastq files can then be processed with <a class="reference external" href="../scripts/split_libraries_fastq.html">split_libraries_fastq.py</a>. Determine which of the following file types you have, and call the corresponding command.</p>
<p>Example iseq with barcode in sequence (more common):</p>
<div class="highlight-python"><div class="highlight"><pre>HWI-ST753_50:6:1101:15435:9071#0/1:ACCAGACGATGCTACGGAGGGAGCTAGCGTTGTTCGGAATTACTGGGCGTAAAGCGCACGTAGGCGGCTTTGTAAGTTAGAGGTGAAAGCCTGGAGCTCAAC:gggggggfggdegggggggggggggggggggegggggggggegggggggeggcccccFUZSU_]]^^ggggggdggdgeeeccYacadcbeddceegggeeg
HWI-ST753_50:6:1101:15446:9128#0/1:AGCTTAACAGCTTACGTAGGGGGCAAGCGTTATCCGGAATTACTGGGTGTAAAGGGAGCGCAGACGGAGAGGCAAGTCAGCTGTGAAAACTCCAGGCTTAAC:BBBBBBBBBBBB`_```_I^HM^`__`____I^^_`_`N``_______`__`___`_\_`G_^L^^^FDJTI^^^ZW^G^BBBBBBBBBBBBBBBBBBBBBB
HWI-ST753_50:6:1101:15300:9134#0/1:ACCAGACGATGCTACGTAGGGGGCAAGCGTTGTCCGGATTTATTGGGCGTAAAGCGCGTGTAGGCGGCCAGGTAGGTCCGTTGTGAAAACTGGAGGCTTAAC:gggggggggcgcggggegggggeggfgggggggggggggggfggggggggggffMffa^cbbgggggggeggdedfb`dfeee`db^fffffge\geggdfg
</pre></div>
</div>
<p>To generate fastq from iseq files with tweleve base barcodes contained as the first bases of the sequence, call the following command:</p>
<div class="highlight-python"><div class="highlight"><pre>process_iseq.py -i s_6_1_sequences.txt,s_7_1_sequences.txt -o ./fastq/ -b 12
</pre></div>
</div>
<p>Example iseq with barcode in header (less common):</p>
<div class="highlight-python"><div class="highlight"><pre>HWI-6X_9267:1:1:12:410#ACAGCTA/1:TACGTAGGGTGCGAGCGTTAATCGGAATTACTGGGCGTAAAGCGTGCGCAGGCGGCATTTTAAGCCAGACGTGAAATCCCCGGGCTTAACCTGGGAACTG:abbb`aaa`^aa```ba`aaaabaaaabaaaa^[Y]^__a`abb`aaaa]Y\\_a[Y_a`a```a__]aaXT\`^\_]`a^^WSZ\JNY]^a`ORO^^`Y
HWI-6X_9267:1:1:12:1762#ACATGAT/1:GACGGAGGATGCAAGTGTTATCCGGAATCACTGGGCGTAAAGCGTCTGTAGGTTGTTTGATAAGTCAACTGTTAAATCTTGAAGCTCAACTTCAAAATCG:aaaaaaaaabaaaaa_aaaaaa`aaaaaaaa`aa``a]aa```a^a^`\```\a`^aaa_\__]]_a_``^``a^^a^b[`SJN]Y_ZZ]^W___`_^U[
HWI-6X_9267:1:1:12:1872#ACAGTTG/1:TACGGAGGGGGTTAGCGTTGTTCCGAATTACTGGGCGTAAAGCGCGCGTAGGCGGATTAGAAAGTTGGGGGGGAAATCCCGGGGCTCAACCCCGGACGTG:aaaaa_aaaa`[a_a`aaaa]a[MY``a\a`aaaaa_\]_\__[_]W]^[[U]aXRZ\W[J\KVTX]\YZZDVY]SUBBBBBBBBBBBBBBBBBBBBBBB
</pre></div>
</div>
<p>To generate fastq from iseq files with six base barcodes contained in the index field of the header, call the following command:</p>
<div class="highlight-python"><div class="highlight"><pre>process_iseq.py -i s_6_1_sequences.txt,s_7_1_sequences.txt -o ./fastq/ --barcode_length 6 --barcode_in_header
</pre></div>
</div>
<p>Note that in the second example there are actually seven bases in the index field. If only six correspond to your barcode (and the remaining bases are technical artifact, for example) you can specify &#8211;barcode_length 6 (as done here) to extract only the first six bases of the barcode.</p>
<p>Once these steps are complete you&#8217;ll have fastq files that can be passed to <a class="reference external" href="../scripts/split_libraries_fastq.html">split_libraries_fastq.py</a>.</p>
</div>
</div>
<div class="section" id="processing-paired-end-read-data-with-qiime">
<h2>Processing paired-end read data with QIIME<a class="headerlink" href="#processing-paired-end-read-data-with-qiime" title="Permalink to this headline">¶</a></h2>
<p>QIIME can be used to process single-end or paired-end read data from the Illumina platform. The primary script for merging paired-end read data in QIIME is <a class="reference external" href="../scripts/join_paired_ends.html">join_paired_ends.py</a>. See the script documentation for more details. This is typically applied as a pre-processing step before running <a class="reference external" href="../scripts/split_libraries_fastq.html">split_libraries_fastq.py</a>.</p>
</div>
<div class="section" id="processing-lea-seq-data-with-qiime">
<h2>Processing LEA-Seq data with QIIME<a class="headerlink" href="#processing-lea-seq-data-with-qiime" title="Permalink to this headline">¶</a></h2>
<p>QIIME provides <strong>beta support</strong> for demultiplexing of sequences using the LEA-Seq protocol described in <a class="reference external" href="http://www.sciencemag.org/content/341/6141/1237439">Faith et al. (2013)</a>. This can be performed using <a class="reference external" href="../scripts/split_libraries_lea_seq.html">split_libraries_lea_seq.py</a>.</p>
</div>
<div class="section" id="working-with-already-demultiplexed-fastq-files">
<h2>Working with already-demultiplexed fastq files<a class="headerlink" href="#working-with-already-demultiplexed-fastq-files" title="Permalink to this headline">¶</a></h2>
<p>Sequencing centers will often perform demultiplexing of sequences for you, delivering you one <tt class="docutils literal"><span class="pre">fastq</span></tt> file per sample. Several scripts are provided that facilitate using these files with QIIME:</p>
<blockquote>
<div><ul class="simple">
<li><a class="reference external" href="../scripts/multiple_split_libraries_fastq.html">multiple_split_libraries_fastq.py</a>: supports quality filtering of already-demultiplexed reads provided in multiple <tt class="docutils literal"><span class="pre">fastq</span></tt> files.</li>
<li><a class="reference external" href="../scripts/multiple_join_paired_ends.html">multiple_join_paired_ends.py</a>: supports joining paired ends for already-demultiplexed reads provided in multiple <tt class="docutils literal"><span class="pre">fastq</span></tt> files.</li>
<li><a class="reference external" href="../scripts/multiple_extract_barcodes.html">multiple_extract_barcodes.py</a>: supports removing barcode sequences from already-demultiplexed reads provided in multiple <tt class="docutils literal"><span class="pre">fastq</span></tt> files. This is useful when sequence barcodes were not removed as part of demultiplexing.</li>
</ul>
</div></blockquote>
</div>
<div class="section" id="working-with-alternate-barcoding-strategies">
<h2>Working with alternate barcoding strategies<a class="headerlink" href="#working-with-alternate-barcoding-strategies" title="Permalink to this headline">¶</a></h2>
<p>In some cases, sequence barcodes are not provided in a separate file, or a dual barcoding strategy may have been applied during sequencing. QIIME supports working with these data by converting these formats into one that is compatible with <a class="reference external" href="../scripts/split_libraries_fastq.html">split_libraries_fastq.py</a>. The examples below illustrate how to use <a class="reference external" href="../scripts/extract_barcodes.html">extract_barcodes.py</a> to accomplish this.</p>
<div class="section" id="processing-a-single-fastq-file-that-starts-with-the-barcode-sequence">
<h3>Processing a single fastq file that starts with the barcode sequence<a class="headerlink" href="#processing-a-single-fastq-file-that-starts-with-the-barcode-sequence" title="Permalink to this headline">¶</a></h3>
<p>In this example, there is a single fastq file, and the first 10 base pairs of the fastq file (in_seqs.fastq) are the barcode</p>
<p>Example of raw data for the file in_seqs.fastq:</p>
<div class="highlight-python"><div class="highlight"><pre>@M10_68:1:1:28680:29475#0/1
AACGAAAGGCAGTTTTGGAAGTAGGCGAATTAGGGTAACGCATATAGGATGCTAATACAACGTGAATGAAGTACTGCATCTATGTCACCAGCTTATTACAGCAGCTTGTCATACATGGCCGTACAGGAAACACACATCATAGCATCACACGA
+
BBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB
@M10_68:1:1:19607:29475#0/1
GACATAAGGGTGGTTAGTATACCGGCAAGGACGGGGTTACTAGTGACGTCCTTCCCCGTATGCCGGGCAATAATGTTTATGTTGGTTTCATGGTTTGGTCTAACTTTACCGCTACTAAATGCTGCGGATTGGTTTCGCTGAATCAGATTATT
+
Z__c\JQ`cc[[_[bfff[[`Qbdge_YYOOHO^cF[FUb_VHMHV`T`dBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB
@M10_68:1:1:22962:29475#0/1
TAATCGAGCTCAACGCCCTGCATACGAAAAGACAGAATCTCTTGCAAGATGTTGGTGCGGTTAGCCAGCTGCTTATGGAAGCCAAGCATTGGGGATTGAGAAAGAGTAGAAATGCCACAAGCCTCAATAGCAGGTTTAAGAGCCTCGATACG
+
JJY````JO[`bab`b`bbaaaaa`\`a`OVT``]]`aa^aI\HMMMWWHHNNNGLL\`________\Z^]]^^^^^^GX]\QTXXZ[YZ^^XZ[Z^\Z^GW\^^\\^^^VZ\Y^^^^\\\\[^[\\\^VWYWWXWWZYZW^[X^\\Z^[TQ
</pre></div>
</div>
<p>An example command to parse out the barcodes and the reads (with barcodes removed) to the output directory parsed_barcodes follows:</p>
<div class="highlight-python"><div class="highlight"><pre>extract_barcodes.py -f in_seqs.fastq --bc1_len 10 -o parsed_barcodes/ --input_type barcode_single_end
</pre></div>
</div>
<p>In the output directory, there should be a <tt class="docutils literal"><span class="pre">barcodes.fastq</span></tt> file with matching labels to the above file that contains the first 10 bases of the reads and the corresponding quality scores, and a <tt class="docutils literal"><span class="pre">reads.fastq</span></tt> containing the remaining bases and quality scores.</p>
<p>If one wanted to reverse complement the barcode before writing it, this command could be used:</p>
<div class="highlight-python"><div class="highlight"><pre>extract_barcodes.py -f in_seqs.fastq --bc1_len 10 --rev_comp_bc1 -o parsed_barcodes/ --input_type barcode_single_end
</pre></div>
</div>
</div>
<div class="section" id="processing-two-fastq-files-that-each-start-with-part-of-a-barcode">
<h3>Processing two fastq files that each start with part of a barcode<a class="headerlink" href="#processing-two-fastq-files-that-each-start-with-part-of-a-barcode" title="Permalink to this headline">¶</a></h3>
<p>In this example, there are two reads, reads1.fastq and reads2.fastq. reads1.fastq starts with 6 bases of a barcode, and reads2.fastq has the remaining 8 bases of this barcode, which need to be pulled from the sequences and combined in an output <tt class="docutils literal"><span class="pre">barcodes.fastq</span></tt> file:</p>
<div class="highlight-python"><div class="highlight"><pre>extract_barcodes.py --input_type barcode_paired_end -f reads1.fastq -r reads2.fastq --bc1_len 6 --bc2_len 8 -o parsed_barcodes/
</pre></div>
</div>
<p>To change the order of the barcodes written, simply change the values/files of the <tt class="docutils literal"><span class="pre">-f</span></tt>, <tt class="docutils literal"><span class="pre">-r</span></tt>, <tt class="docutils literal"><span class="pre">--bc1_len</span></tt>, and <tt class="docutils literal"><span class="pre">--bc2_len</span></tt> parameters:</p>
<div class="highlight-python"><div class="highlight"><pre>extract_barcodes.py --input_type barcode_paired_end -f reads2.fastq -r reads1.fastq --bc1_len 8 --bc2_len 6 -o parsed_barcodes/
</pre></div>
</div>
<p>The barcodes can each be reverse complemented before writing via the <tt class="docutils literal"><span class="pre">--rev_comp_bc1</span></tt> and <tt class="docutils literal"><span class="pre">--rev_comp_bc2</span></tt> parameters.</p>
<p>In some sequencing reactions, the orientation of the reads is random, so in the reads1.fastq and reads2.fastq example above, there would be a mixture of amplicons in different orientations. One solution is to attempt to orient the reads by looking for the forward and reverse primers in the sequences, and selectively writing out reads that match the orientation of the forward primer to reads1 and reads that match the reverse primer to reads2. One must supply a metadata mapping file that contains the LinkerPrimerSequence and ReversePrimer fields (all primers should be written in 5&#8217;-&gt;3&#8217; orientation). See <a class="reference external" href="../documentation/file_formats.html#metadata-mapping-files">metadata mapping file format</a> for more information. An example command to attempt read orientation:</p>
<div class="highlight-python"><div class="highlight"><pre>extract_barcodes.py --input_type barcode_paired_end -m mapping_file.txt -a -f reads1.fastq -r reads2.fastq --bc1_len 6 --bc2_len 8 -o parsed_barcodes/
</pre></div>
</div>
<p>In addition to the normal <tt class="docutils literal"><span class="pre">barcodes.fastq</span></tt>, <tt class="docutils literal"><span class="pre">reads1.fastq</span></tt>, and <tt class="docutils literal"><span class="pre">reads2.fastq</span></tt> files, there will be additional files labeled as <tt class="docutils literal"><span class="pre">_not_oriented.fastq</span></tt> where the primers could not be found. These are all written and have the barcodes processed in the order that the files were supplied in.</p>
<p>With paired sequences, one could run into issues of labels not matching-this can be suppressed with the <tt class="docutils literal"><span class="pre">--disable_header_match</span></tt> option, but one should be careful about making sure the reads are actually matched up if this option is used.</p>
</div>
<div class="section" id="two-index-barcode-reads-and-two-fastq-reads">
<h3>Two index/barcode reads and two fastq reads<a class="headerlink" href="#two-index-barcode-reads-and-two-fastq-reads" title="Permalink to this headline">¶</a></h3>
<p>This situation can be treated as a special case of paired-end reads. One could supply the index files (labeled as index1.fastq, index2.fastq) and use the <tt class="docutils literal"><span class="pre">--input_type</span> <span class="pre">barcode_paired_end</span></tt>:</p>
<div class="highlight-python"><div class="highlight"><pre>extract_barcodes.py --input_type barcode_paired_end -f index1.fastq -r index2.fastq --bc1_len 6 --bc2_len 6 -o parsed_barcodes/
</pre></div>
</div>
<p>The output <tt class="docutils literal"><span class="pre">barcodes.fastq</span></tt> file would be used for downstream processing, and the reads1 and reads2 files could be ignored.</p>
</div>
<div class="section" id="processing-a-single-stitched-read-that-contains-barcodes-on-each-end">
<h3>Processing a single stitched read that contains barcodes on each end<a class="headerlink" href="#processing-a-single-stitched-read-that-contains-barcodes-on-each-end" title="Permalink to this headline">¶</a></h3>
<p>In this case, a single stitched fastq file (reads.fastq) is present that contains 6 base pair barcodes on each end. We can write the barcodes and the stitched read (sans barcodes) to the output directory parsed_barcodes with this command:</p>
<div class="highlight-python"><div class="highlight"><pre>extract_barcodes.py --input_type barcode_paired_stitched -f reads.fastq --bc1_len 6 --bc2_len 6 -o parsed_barcodes/
</pre></div>
</div>
<p>To switch the order of the barcodes (i.e., the barcode at the end of the read is written first), the <tt class="docutils literal"><span class="pre">--switch_bc_order</span></tt> option could be enabled. The barcodes can individually, or both, be reverse complemented in the same manner as described above. Additionally, one can attempt to orient the sequences as described in the paired-end read example. In this case, the orientation process will reverse complement the stitched reads to try and match the orientation of the forward primers.</p>
</div>
<div class="section" id="barcodes-in-fastq-labels">
<h3>Barcodes in fastq labels<a class="headerlink" href="#barcodes-in-fastq-labels" title="Permalink to this headline">¶</a></h3>
<p>In some cases, sequencing centers will put the barcode sequences in the fastq labels before handing it off. It is important to look for the character immediately in front of the barcode and the length of the barcode:</p>
<div class="highlight-python"><div class="highlight"><pre>@MCIC-SOLEXA_0051_FC:1:1:4065:1039#CGATGT/1
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
+MCIC-SOLEXA_0051_FC:1:1:4065:1039#CGATGT/1
KPPPQWWWWWQQ________BBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB
</pre></div>
</div>
<p>In this case, the &#8220;#&#8221; character is before the barcode, and the barcodes are 6 base pairs in length. To parse this example file, called in_seqs.fastq, this example command could be used:</p>
<div class="highlight-python"><div class="highlight"><pre>extract_barcodes.py --input_type barcode_in_label --char_delineator &quot;#&quot; -f in_seqs.fastq --bc1_len 6 -o parsed_barcodes/
</pre></div>
</div>
<p>A second fastq file could be passed (<tt class="docutils literal"><span class="pre">-r</span></tt>) if one had paired files with barcodes in the labels, and the parameters for changing barcode lengths or reverse complementing barcodes all apply.</p>
</div>
</div>
<div class="section" id="processing-joint-genome-institute-fastq-files">
<h2>Processing Joint Genome Institute fastq files<a class="headerlink" href="#processing-joint-genome-institute-fastq-files" title="Permalink to this headline">¶</a></h2>
<p>The Joint Genome Institute (JGI) delivers a single fastq file to their users, which is not compatible with <a class="reference external" href="../scripts/split_libraries_fastq.html">split_libraries_fastq.py</a>. To fix this issue, first we can split this file into forward and reverse reads and then extract the barcode from the forward file.</p>
<p>The fastq created by JGI alternates the forward and reverse reads in a single file and has the barcode in the header of the file, for example:</p>
<div class="highlight-python"><div class="highlight"><pre>@MISEQ03:64:000000000-A2H3D:1:1101:14358:1530 1:N:0:TCCACAGGAGT
TNCAGAGGGTGCAAGCGTTAATCGGAATTACTGGGCGTAAAGCGCGCGTAGGTGGTTTGTTAAGTTGGATGTGAAATCCCCGGGCTCAACCTGGGAACTGCATTCAAAACTGACAAGCTAGAGTATGGTAGAGGGTGGTGGAATTTCCTGTGTAGCGGTGAAATGCGTAGATATAGGAAGGAACACCAGTGGCGAAGGCGACCACCTGGACTGAAACTGACACTGAGGGGCGAAAGCGGGGGGGGCAAACG
+
?#5&lt;????DDDDDDDDEEEEFFHHHHHHHHHHHHHHDCCHHFGDEHEH&gt;CCE5AEEHHHHHHHHHHHHHHHHHFFFFHHHHHHEEADEEEEEEEEEEEEEEEEEEEEEEE?BEEEEEEEEEEEAEEEE0?A:?EE)8;)0ACEEECECCECAACEE?&gt;)8CCC?CCA8?88ACC*A*::A??:0?C?.?0:?8884&gt;&#39;.&#39;&#39;..&#39;0?8C?C**0:0::?ECEE?############################
@MISEQ03:64:000000000-A2H3D:1:1101:14358:1530 2:N:0:TCCACAGGAGT
ACGGACTACAAGGGTTTCTAATCCTGTTTGCTCCCCACGCTTTCGCACCTCAGTGTCAGTATCAGTCCAGGTGGTCGCCTTCGCCACTGGTGTTCCTTCCTATATCTACGCATTTCACCGCTACACAGGAAATTCCACCACCCTCTACCATACTCTAGCTTGTCAGTTTTGAATGCAGTTCCCAGGTTGAGCCCGGGGATTTCACATCCAACTTAACAACCCACATACCCGCCTTTTCGCCCAGGTAATCC
+
?????@@BDDBDDDBBCFFCFFHIIIIIIIIFGHHHHEHHHIIIHHHHHFHIIHIGHHIDGGHHHHIIIIICEFHIHHCDEHHHHHHFHHCFHDF?FHHFHHHFFDFFFDEDDD..=DDDE@&lt;BFEEFCFFCECE==CACFE?*0:*CCAA?:*:*:0*A?A80:???A?*00:**0*1*:C??C?A?01*0?);&gt;&gt;&#39;.8::A?###############################################
@MISEQ03:64:000000000-A2H3D:1:1101:14206:1564 1:N:0:TCCACAGGAGT
TACGTAGGGTGCGAGCGTTAATCGGAATTACTGGGCGTAAAGCGTGCGCAGGCGGTTTTGTAAGTCAGATGTGAAAGCCCCGGGCTCAACCTGGGAACTGCGTTTGAAACTACAAGGCTAGAGTGTAGCAGAGGGGGGTAGAATTCCACGTGTAGCGGTGAAATGCGTAGAGATGGGGAGGAATACCAATGGCGAAGGCAGCCCCCGGGGTTAACACTGACGCCAAGGCACGAAAGCGGGGGGGGCAAACG
+
?????BB?DDDDDD@DDCEEFFH&gt;EEFHHHHHHGHHHCEEFFDC5EECCCCCCDECEHF;?DFDDFHDDBBDF?CFDCCFEA@@::;EEEEEEEECBA,BBE?:&gt;AA?CA*:**0:??A:8*:*0*0**0*:?CE?DD&#39;..0????:*:?*0?EC*&#39;.)4.?A***00)&#39;.00*0*08)8??8*0:CEE*0:082.4;**?AEAA?#############################################
@MISEQ03:64:000000000-A2H3D:1:1101:14206:1564 2:N:0:TCCACAGGAGT
ACGGACTACAGGGGTTTCTAATCCTGTTTGCTCCCCACGCTTTCGTGCATGAGCGTCAGTGTTAACCCAGGGGGCTGCCTTCGCCATTGGTATTCCTCCACATCTCTACGCATTTCACTGCTACACGTGGAATTCTACCCCCCTCTGCTACACTCTAGCCTTGTAGTTTCAAACGCAGTTCCCAGGTTGAGCCCGGGGCTTTCACATCTGCCTTACAAAACCGCCTGCGCACGCTTTACGCCCCGTAATTC
+
?????@@BDDBDDD?@CFFFFFHHHHHFFHHHHHHHHHHH@FFHEDFFH@FHBGCDHHHBFHHHHHHHEHHHHDCCEFFDFFFEE:=?FF?DFDFDFFF==BEE=DBDDEEEEEB,4??EE@EEE,3,3*3,?:?*0ACCEDD88:***?*0:*0***0*?C?00:AE:?EE:*A8&#39;.?:CAA?A80*0*??AA88;28;C##################################################
</pre></div>
</div>
<p>To split the input file in forward and reverse we will use <a class="reference external" href="../scripts/extract_reads_from_interleaved_file.html">extract_reads_from_interleaved_file.py</a>:</p>
<div class="highlight-python"><div class="highlight"><pre>extract_reads_from_interleaved_file.py -i reads_to_extract.fastq -o extracted_reads
</pre></div>
</div>
<p>Then the next step is to create the barcode (index) fastq file by using the forward read generated in the previous step and <a class="reference external" href="../scripts/extract_barcodes.html">extract_barcodes.py</a>. Note how we are using the parameter <tt class="docutils literal"><span class="pre">-l</span> <span class="pre">11</span></tt> because the primers in this example have that length, see header in the example:</p>
<div class="highlight-python"><div class="highlight"><pre>extract_barcodes.py -f extracted_reads/forward_reads.fastq -o barcode_reads -c barcode_in_label --char_delineator &#39;1:N:0:&#39; -l 11
</pre></div>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">

    
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Preparing raw Illumina data in different formats for use with QIIME</a><ul>
<li><a class="reference internal" href="#quality-filtering-of-illumina-data-with-qiime">Quality filtering of Illumina data with QIIME</a></li>
<li><a class="reference internal" href="#demultiplexing-and-quality-filtering-fastq-illumina-data-with-qiime">Demultiplexing and quality filtering fastq Illumina data with QIIME</a><ul>
<li><a class="reference internal" href="#starting-with-gzipped-fastq-files">Starting with gzipped fastq files</a></li>
<li><a class="reference internal" href="#processing-multiple-lanes-of-illumina-fastq-data-with-qiime">Processing multiple lanes of Illumina fastq data with QIIME</a></li>
</ul>
</li>
<li><a class="reference internal" href="#demultiplexing-and-quality-filtering-non-fastq-illumina-data-with-qiime">Demultiplexing and quality filtering non-fastq Illumina data with QIIME</a><ul>
<li><a class="reference internal" href="#processing-illumina-qseq-files">Processing Illumina qseq files</a></li>
<li><a class="reference internal" href="#processing-iseq-files-with-qiime">Processing iseq files with QIIME</a></li>
</ul>
</li>
<li><a class="reference internal" href="#processing-paired-end-read-data-with-qiime">Processing paired-end read data with QIIME</a></li>
<li><a class="reference internal" href="#processing-lea-seq-data-with-qiime">Processing LEA-Seq data with QIIME</a></li>
<li><a class="reference internal" href="#working-with-already-demultiplexed-fastq-files">Working with already-demultiplexed fastq files</a></li>
<li><a class="reference internal" href="#working-with-alternate-barcoding-strategies">Working with alternate barcoding strategies</a><ul>
<li><a class="reference internal" href="#processing-a-single-fastq-file-that-starts-with-the-barcode-sequence">Processing a single fastq file that starts with the barcode sequence</a></li>
<li><a class="reference internal" href="#processing-two-fastq-files-that-each-start-with-part-of-a-barcode">Processing two fastq files that each start with part of a barcode</a></li>
<li><a class="reference internal" href="#two-index-barcode-reads-and-two-fastq-reads">Two index/barcode reads and two fastq reads</a></li>
<li><a class="reference internal" href="#processing-a-single-stitched-read-that-contains-barcodes-on-each-end">Processing a single stitched read that contains barcodes on each end</a></li>
<li><a class="reference internal" href="#barcodes-in-fastq-labels">Barcodes in fastq labels</a></li>
</ul>
</li>
<li><a class="reference internal" href="#processing-joint-genome-institute-fastq-files">Processing Joint Genome Institute fastq files</a></li>
</ul>
</li>
</ul>


    <h3><a href="../index.html">Site index</a></h3>
    <ul><li><ul>
    <li><a href="http://www.qiime.org">Home</a><br /></li>
    <li><a href="../install/index.html">Install</a><br /></li>
    <li><a href="index.html">Tutorials</a><br /></li>
    <li><a href="../scripts/index.html">Scripts</a><br /></li>
    <li><a href="http://help.qiime.org">Help</a><br /></li>
    <li><a href="http://qiime.org/home_static/dataFiles.html">Resources</a><br /></li>
    <li><a href="../documentation/index.html">File Formats</a><br /></li>
    <li><a href="http://workshops.qiime.org">Workshops</a><br /></li>
    <li><a href="http://qiime.wordpress.com">Blog</a><br /></li>
    <li><a href="../developer/index.html">Developer</a><br /></li>
    </ul></li></ul>


  <h4>Previous topic</h4>
  <p class="topless"><a href="fungal_its_analysis.html"
                        title="previous chapter">Fungal ITS analysis tutorial (an IPython Notebook): open-reference OTU picking</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="processing_18S_data.html"
                        title="next chapter">Analysis of 18S data</a></p>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="processing_18S_data.html" title="Analysis of 18S data"
             >next</a> |</li>
        <li class="right" >
          <a href="fungal_its_analysis.html" title="Fungal ITS analysis tutorial (an IPython Notebook): open-reference OTU picking"
             >previous</a> |</li>
        <li><a href="../index.html">Home</a> &raquo;</li>
          <li><a href="index.html" >QIIME Tutorials</a> &raquo;</li> 
      </ul>
    </div>
<div style="text-align: center; padding: 0px;"><br />
<a href="http://www.scikit-bio.org" target="_blank"><img src="../_static/powered_by_skbio.png" border="0" alt="sampledoc" height="70px"/></a>
<br /></div>

    <div class="footer">
        &copy; Copyright 2010-2014, QIIME Team.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.2.3.
    </div>
<script type="text/javascript">
var gaJsHost = (("https:" == document.location.protocol) ? "https://ssl." : "http://www.");
document.write(unescape("%3Cscript src='" + gaJsHost + "google-analytics.com/ga.js' type='text/javascript'%3E%3C/script%3E"));
</script>
<script type="text/javascript">
try {
var pageTracker = _gat._getTracker("UA-6636235-4");
pageTracker._trackPageview();
} catch(err) {}</script>

  </body>
</html>