<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Discussion of subsampled open-reference OTU picking in QIIME &mdash; Homepage</title>
    
    <link rel="stylesheet" href="../_static/default.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '1.9.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <link rel="shortcut icon" href="../_static/favicon.ico"/>
    <link rel="top" title="Homepage" href="../index.html" />
    <link rel="up" title="QIIME Tutorials" href="index.html" />
    <link rel="next" title="Fungal ITS analysis tutorial (an IPython Notebook): open-reference OTU picking" href="fungal_its_analysis.html" />
    <link rel="prev" title="454 Overview Tutorial: de novo OTU picking and diversity analyses using 454 data" href="tutorial.html" />
<meta http-equiv="Content-Style-Type" content="text/css" />
<script type="text/javascript" src="http://www.google.com/jsapi?key=ABQIAAAAbW_pA971hrPgosv-Msv7hRRE2viNBUPuU405tK6p2cguOFmlFBQSwZMG6_q_v6Z42nkdo9ejT1aHmA"></script>
<script type="text/javascript" src="../_static/google_feed.js"></script>

  </head>
  <body>

<div style="background-color: white; text-align: left; padding: 0px">
<a href="../index.html"><img src="../_static/wordpressheader.png" border="0" alt="sampledoc"/></a>
</div>
<div class="news" style="background-color:#ccc;"><table style="font-size:12pt;padding: 5px;"><tr id="feed"><td><b>News and Announcements &raquo;</b></td>
</tr></table></div>

    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="fungal_its_analysis.html" title="Fungal ITS analysis tutorial (an IPython Notebook): open-reference OTU picking"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="tutorial.html" title="454 Overview Tutorial: de novo OTU picking and diversity analyses using 454 data"
             accesskey="P">previous</a> |</li>
        <li><a href="../index.html">Home</a> &raquo;</li>
          <li><a href="index.html" accesskey="U">QIIME Tutorials</a> &raquo;</li> 
      </ul>
    </div>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="discussion-of-subsampled-open-reference-otu-picking-in-qiime">
<span id="open-reference-illumina"></span><h1>Discussion of subsampled open-reference OTU picking in QIIME<a class="headerlink" href="#discussion-of-subsampled-open-reference-otu-picking-in-qiime" title="Permalink to this headline">¶</a></h1>
<p>For a definition and detailed discussion of the <em>subsampled open-reference OTU picking protocol</em> (and comparison to other OTU picking protocols), please see <a class="reference external" href="https://peerj.com/articles/545/">Rideout et al. (2014)</a>. This tutorial supplements this paper by describing how to use QIIME&#8217;s subsampled open-reference OTU picking workflow.</p>
<blockquote>
<div><div class="toctree-wrapper compound">
<ul class="simple">
</ul>
</div>
</div></blockquote>
<div class="section" id="subsampled-open-reference-otu-picking">
<h2>Subsampled open-reference OTU picking<a class="headerlink" href="#subsampled-open-reference-otu-picking" title="Permalink to this headline">¶</a></h2>
<p>Subsampled reference OTU picking is suitable for any analysis that legacy open-reference OTU picking can be used for, but additionally scales to much larger data sets (such as multiple HiSeq runs, which may require several days on ~100 processors to analyze).</p>
<blockquote>
<div><div class="admonition warning">
<p class="first admonition-title">Warning</p>
<p class="last">If processing multiple HiSeq lanes, don&#8217;t combine the sequence data into a single file. Instead, see <a class="reference internal" href="#iterative-mode"><em>Using the subsampled open-reference OTU picking workflow in iterative mode</em></a>.</p>
</div>
</div></blockquote>
<p>This is an open-reference OTU picking protocol, meaning that sequences are clustered against a reference database, and reads which fail to hit the reference are subsequently clustered de novo. This differs from legacy open-reference OTU picking as it was optimized at several steps to enable running on massive numbers of sequences (e.g., hundreds of millions). The steps in this workflow are as follows.</p>
<div class="section" id="step-0-optional-prefilter-parallel">
<h3>Step 0: Optional prefilter (parallel)<a class="headerlink" href="#step-0-optional-prefilter-parallel" title="Permalink to this headline">¶</a></h3>
<p>Prefilter the input sequence collection by searching reads against the reference set with a low percent identity threshold to discard sequences that are not representatives of the targeted marker gene (e.g., sequences that are the result of sequencing error, or host genomic sequences). This filter is disabled by default as it is very slow to run, and the same sequences are often discarded later in the process as they fail to align with PyNAST. If you&#8217;d like to use this filter and are working with 16S sequence data, a threshold of 60% is a good place to start (passed as <tt class="docutils literal"><span class="pre">--prefilter_percent_id</span></tt>). The choice of 60% is described in <a class="reference external" href="https://peerj.com/articles/545/">Rideout et al. (2014)</a>. All reads which fail to hit the reference set are discarded as likely sequencing error.</p>
<blockquote>
<div><div class="admonition warning">
<p class="first admonition-title">Warning</p>
<p class="last">If most or all of your sequences are being filtered at this step, your sequences may be in the reverse orientation with respect to your reference database. To address this, you should add the following to your parameters file (creating one, if necessary) and pass this file as <tt class="docutils literal"><span class="pre">-p</span></tt> to <tt class="docutils literal"><span class="pre">pick_open_reference_otus.py</span></tt>: <tt class="docutils literal"><span class="pre">pick_otus:enable_rev_strand_match</span> <span class="pre">True</span></tt>. This is included in the instructions below, but be aware that this doubles the memory used in this step of the workflow.</p>
</div>
</div></blockquote>
</div>
<div class="section" id="step-1-closed-reference-parallel">
<h3>Step 1: Closed reference (parallel)<a class="headerlink" href="#step-1-closed-reference-parallel" title="Permalink to this headline">¶</a></h3>
<p>Apply closed-reference OTU picking against the reference collection. Generate a fasta file containing all reads that fail to hit the reference collection.</p>
</div>
<div class="section" id="step-2-de-novo-clustering-of-subsampled-failures-serial">
<h3>Step 2: De novo clustering of subsampled failures (serial)<a class="headerlink" href="#step-2-de-novo-clustering-of-subsampled-failures-serial" title="Permalink to this headline">¶</a></h3>
<p>Randomly subsample the sequences that failed to hit the reference collection, and write these to a new fasta file (default subsampling percentage is 0.1%, modify with <tt class="docutils literal"><span class="pre">-s/--percent_subsample</span></tt>). Cluster these reads de novo, and choose a representative set of sequences as the centroid of each OTU cluster. These are the <em>new reference</em> OTUs.</p>
</div>
<div class="section" id="step-3-closed-reference-round-2-parallel">
<h3>Step 3: Closed reference, round 2 (parallel)<a class="headerlink" href="#step-3-closed-reference-round-2-parallel" title="Permalink to this headline">¶</a></h3>
<p>Pick closed-reference OTUs against the representative sequences from the previous step. Write all sequences that fail to hit the reference collection to a fasta file.</p>
</div>
<div class="section" id="step-4-de-novo-serial">
<h3>Step 4: De novo (serial)<a class="headerlink" href="#step-4-de-novo-serial" title="Permalink to this headline">¶</a></h3>
<p>Pick de novo OTUs on all reads that failed to hit the reference collection in the previous step. These are the <em>clean-up</em> OTUs. This step can be suppress by passing <tt class="docutils literal"><span class="pre">--suppress_step4</span></tt>.</p>
</div>
<div class="section" id="post-otu-processing-parallel-and-serial-depending-on-step">
<h3>Post-OTU processing (parallel and serial, depending on step)<a class="headerlink" href="#post-otu-processing-parallel-and-serial-depending-on-step" title="Permalink to this headline">¶</a></h3>
<ol class="arabic simple">
<li>Assemble the reference OTUs, the new reference OTUs, and the clean-up OTUs into a new OTU map, and construct an OTU table. At this stage, all OTUs with a sequence count of smaller than 2 (i.e., the singleton OTUs) are discarded. This can be modified with the <tt class="docutils literal"><span class="pre">--min_otu_size</span></tt> option, and disabled by passing <tt class="docutils literal"><span class="pre">--min_otu_size=1</span></tt>.</li>
<li>Construct a new reference collection based on this OTU picking run. This new reference collection will be the combination of the full input reference collection, the new reference OTU representative sequences, and the clean-up OTU representative sequences. Note that this will not include representatives of the singleton OTUs by default. Also note that this differs from the representative set of sequences for this run in that it contains <em>all</em> of the input reference sequences, not only the ones that are represented in the current data set (which is what the representative sequences for this run contains).</li>
<li>Taxonomy will be assigned to all OTUs (using the uclust consensus taxonomy assigner by default) and representative sequences will be aligned and a tree will be constructed. Finally, an additional OTU table will be constructed that excludes reads that failed to align with PyNAST. It is recommended that this OTU table be used in downstream analysis.</li>
</ol>
<p>To apply this analysis to <tt class="docutils literal"><span class="pre">seqs1.fna</span></tt>, picking OTUs against the reference collection <tt class="docutils literal"><span class="pre">refseqs.fna</span></tt> you can run the following command.</p>
<p>You should <em>always use full paths</em> which are represented here by <tt class="docutils literal"><span class="pre">$PWD</span></tt>, but will usually look something like <tt class="docutils literal"><span class="pre">$HOME/my_data/</span></tt> (in other words, they should start with a <tt class="docutils literal"><span class="pre">/</span></tt>). In this example your input and reference sequences (<tt class="docutils literal"><span class="pre">seqs1.fna</span></tt> and <tt class="docutils literal"><span class="pre">refseqs.fna</span></tt>) are in the directory represented by <tt class="docutils literal"><span class="pre">$PWD</span></tt>. If you work from the directory containing those files, you can leave <tt class="docutils literal"><span class="pre">$PWD</span></tt> in the commands instead of specifying the full paths:</p>
<div class="highlight-python"><div class="highlight"><pre>pick_open_reference_otus.py -i $PWD/seqs1.fna -r $PWD/refseqs.fna -o $PWD/ucrss/ -aO 8
</pre></div>
</div>
<p>This command should be run in parallel. Each job will need approximately 4GB of RAM, so if running on EC2 and you want to start 8 parallel jobs (recommended setting for EC2), your instance type should be <tt class="docutils literal"><span class="pre">m2.4xlarge</span></tt>. The <tt class="docutils literal"><span class="pre">-aO</span> <span class="pre">8</span></tt> specifies that we want to start 8 parallel jobs - adjust this according to the resources you have available.</p>
<p>When this job completes, you&#8217;re ready to begin running diversity analyses. Please see the <a class="reference external" href="./illumina_overview_tutorial.html">QIIME Illumina Overview Tutorial</a>, which covers the next steps starting from the output of <tt class="docutils literal"><span class="pre">pick_open_reference_otus.py</span></tt>.</p>
</div>
</div>
<div class="section" id="filtering-an-open-reference-otu-table-to-reference-otus-only">
<span id="filter-to-closed-ref"></span><h2>Filtering an open-reference OTU table to reference OTUs only<a class="headerlink" href="#filtering-an-open-reference-otu-table-to-reference-otus-only" title="Permalink to this headline">¶</a></h2>
<p>There are cases where you may be interested in working with the closed reference subset of your open reference OTU table (meaning only those OTUs that hit the reference collection, excluding the new OTUs). Following from the above commands, to do that you can filter the new OTUs from the OTU table with the following command:</p>
<div class="highlight-python"><div class="highlight"><pre>filter_otus_from_otu_table.py -i $PWD/ucrss/otu_table_mc2_w_tax_no_pynast_failures.biom -o $PWD/ucrss/otu_table_mc2_w_tax_no_pynast_failures.reference_only.biom --negate_ids_to_exclude -e $PWD/refseqs.fna
</pre></div>
</div>
<p>This excludes all OTUs with identifiers that are not present in <tt class="docutils literal"><span class="pre">$PWD/refseqs.fna</span></tt>, thus removing all of the non-reference OTUs.</p>
</div>
<div class="section" id="using-the-subsampled-open-reference-otu-picking-workflow-in-iterative-mode">
<span id="iterative-mode"></span><h2>Using the subsampled open-reference OTU picking workflow in iterative mode<a class="headerlink" href="#using-the-subsampled-open-reference-otu-picking-workflow-in-iterative-mode" title="Permalink to this headline">¶</a></h2>
<p>The subsampled open-reference OTU picking workflow can be run in iterative mode to support multiple different sequence collections, such as several HiSeq runs. In iterative mode, the list of sequence files will be processed in order, and the new reference sequences generated at each step will be used as the reference collection for the subsequent step. After all input collections have been processed, a single OTU table and tree covering all of the input collections will be generated.</p>
<p>To apply this analysis to <tt class="docutils literal"><span class="pre">seqs1.fna</span></tt> and <tt class="docutils literal"><span class="pre">seqs2.fna</span></tt> in iterative mode, picking OTUs against the reference collection <tt class="docutils literal"><span class="pre">refseqs.fna</span></tt> you can run the following command.</p>
<p>You should <em>always use full paths</em> which are represented here by <tt class="docutils literal"><span class="pre">$PWD</span></tt>, but will usually look something like <tt class="docutils literal"><span class="pre">$HOME/my_data/</span></tt> (in other words, they should start with a <tt class="docutils literal"><span class="pre">/</span></tt>). In this example your input and reference sequences (<tt class="docutils literal"><span class="pre">seqs1.fna</span></tt>, <tt class="docutils literal"><span class="pre">seqs2.fna</span></tt>, and <tt class="docutils literal"><span class="pre">refseqs.fna</span></tt>) are in the directory represented by <tt class="docutils literal"><span class="pre">$PWD</span></tt>. If you work from the directory containing those files, you can leave <tt class="docutils literal"><span class="pre">$PWD</span></tt> in the commands instead of specifying the full paths:</p>
<div class="highlight-python"><div class="highlight"><pre>pick_open_reference_otus.py -i $PWD/seqs1.fna,$PWD/seqs2.fna -r $PWD/refseqs.fna -o $PWD/ucrss_iter/ -aO 8
</pre></div>
</div>
<p>This command should be run in parallel. Each job will need approximately 4GB of RAM, so if running on EC2 and you want to start 8 parallel jobs (recommended setting for EC2), your instance type should be <tt class="docutils literal"><span class="pre">m2.4xlarge</span></tt>. The <tt class="docutils literal"><span class="pre">-aO</span> <span class="pre">8</span></tt> specifies that we want to start 8 parallel jobs - adjust this according to the resources you have available.</p>
<p>After iterative OTU picking you can continue on with diversity analyses as described in the <a class="reference external" href="./illumina_overview_tutorial.html">QIIME Illumina Overview Tutorial</a>.</p>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">

    
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Discussion of subsampled open-reference OTU picking in QIIME</a><ul>
<li><a class="reference internal" href="#subsampled-open-reference-otu-picking">Subsampled open-reference OTU picking</a><ul>
<li><a class="reference internal" href="#step-0-optional-prefilter-parallel">Step 0: Optional prefilter (parallel)</a></li>
<li><a class="reference internal" href="#step-1-closed-reference-parallel">Step 1: Closed reference (parallel)</a></li>
<li><a class="reference internal" href="#step-2-de-novo-clustering-of-subsampled-failures-serial">Step 2: De novo clustering of subsampled failures (serial)</a></li>
<li><a class="reference internal" href="#step-3-closed-reference-round-2-parallel">Step 3: Closed reference, round 2 (parallel)</a></li>
<li><a class="reference internal" href="#step-4-de-novo-serial">Step 4: De novo (serial)</a></li>
<li><a class="reference internal" href="#post-otu-processing-parallel-and-serial-depending-on-step">Post-OTU processing (parallel and serial, depending on step)</a></li>
</ul>
</li>
<li><a class="reference internal" href="#filtering-an-open-reference-otu-table-to-reference-otus-only">Filtering an open-reference OTU table to reference OTUs only</a></li>
<li><a class="reference internal" href="#using-the-subsampled-open-reference-otu-picking-workflow-in-iterative-mode">Using the subsampled open-reference OTU picking workflow in iterative mode</a></li>
</ul>
</li>
</ul>


    <h3><a href="../index.html">Site index</a></h3>
    <ul><li><ul>
    <li><a href="http://www.qiime.org">Home</a><br /></li>
    <li><a href="../install/index.html">Install</a><br /></li>
    <li><a href="index.html">Tutorials</a><br /></li>
    <li><a href="../scripts/index.html">Scripts</a><br /></li>
    <li><a href="http://help.qiime.org">Help</a><br /></li>
    <li><a href="http://qiime.org/home_static/dataFiles.html">Resources</a><br /></li>
    <li><a href="../documentation/index.html">File Formats</a><br /></li>
    <li><a href="http://workshops.qiime.org">Workshops</a><br /></li>
    <li><a href="http://qiime.wordpress.com">Blog</a><br /></li>
    <li><a href="../developer/index.html">Developer</a><br /></li>
    </ul></li></ul>


  <h4>Previous topic</h4>
  <p class="topless"><a href="tutorial.html"
                        title="previous chapter">454 Overview Tutorial: de novo OTU picking and diversity analyses using 454 data</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="fungal_its_analysis.html"
                        title="next chapter">Fungal ITS analysis tutorial (an IPython Notebook): open-reference OTU picking</a></p>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="fungal_its_analysis.html" title="Fungal ITS analysis tutorial (an IPython Notebook): open-reference OTU picking"
             >next</a> |</li>
        <li class="right" >
          <a href="tutorial.html" title="454 Overview Tutorial: de novo OTU picking and diversity analyses using 454 data"
             >previous</a> |</li>
        <li><a href="../index.html">Home</a> &raquo;</li>
          <li><a href="index.html" >QIIME Tutorials</a> &raquo;</li> 
      </ul>
    </div>
<div style="text-align: center; padding: 0px;"><br />
<a href="http://www.scikit-bio.org" target="_blank"><img src="../_static/powered_by_skbio.png" border="0" alt="sampledoc" height="70px"/></a>
<br /></div>

    <div class="footer">
        &copy; Copyright 2010-2014, QIIME Team.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.2.3.
    </div>
<script type="text/javascript">
var gaJsHost = (("https:" == document.location.protocol) ? "https://ssl." : "http://www.");
document.write(unescape("%3Cscript src='" + gaJsHost + "google-analytics.com/ga.js' type='text/javascript'%3E%3C/script%3E"));
</script>
<script type="text/javascript">
try {
var pageTracker = _gat._getTracker("UA-6636235-4");
pageTracker._trackPageview();
} catch(err) {}</script>

  </body>
</html>