<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>454 Overview Tutorial: de novo OTU picking and diversity analyses using 454 data &mdash; Homepage</title>
    
    <link rel="stylesheet" href="../_static/default.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '1.9.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <link rel="shortcut icon" href="../_static/favicon.ico"/>
    <link rel="top" title="Homepage" href="../index.html" />
    <link rel="up" title="QIIME Tutorials" href="index.html" />
    <link rel="next" title="Discussion of subsampled open-reference OTU picking in QIIME" href="open_reference_illumina_processing.html" />
    <link rel="prev" title="Illumina Overview Tutorial (an IPython Notebook): open reference OTU picking and core diversity analyses" href="illumina_overview_tutorial.html" />
<meta http-equiv="Content-Style-Type" content="text/css" />
<script type="text/javascript" src="http://www.google.com/jsapi?key=ABQIAAAAbW_pA971hrPgosv-Msv7hRRE2viNBUPuU405tK6p2cguOFmlFBQSwZMG6_q_v6Z42nkdo9ejT1aHmA"></script>
<script type="text/javascript" src="../_static/google_feed.js"></script>

  </head>
  <body>

<div style="background-color: white; text-align: left; padding: 0px">
<a href="../index.html"><img src="../_static/wordpressheader.png" border="0" alt="sampledoc"/></a>
</div>
<div class="news" style="background-color:#ccc;"><table style="font-size:12pt;padding: 5px;"><tr id="feed"><td><b>News and Announcements &raquo;</b></td>
</tr></table></div>

    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="open_reference_illumina_processing.html" title="Discussion of subsampled open-reference OTU picking in QIIME"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="illumina_overview_tutorial.html" title="Illumina Overview Tutorial (an IPython Notebook): open reference OTU picking and core diversity analyses"
             accesskey="P">previous</a> |</li>
        <li><a href="../index.html">Home</a> &raquo;</li>
          <li><a href="index.html" accesskey="U">QIIME Tutorials</a> &raquo;</li> 
      </ul>
    </div>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="overview-tutorial-de-novo-otu-picking-and-diversity-analyses-using-454-data">
<span id="tutorial"></span><h1>454 Overview Tutorial: de novo OTU picking and diversity analyses using 454 data<a class="headerlink" href="#overview-tutorial-de-novo-otu-picking-and-diversity-analyses-using-454-data" title="Permalink to this headline">¶</a></h1>
<p>This tutorial explains how to apply de novo OTU picking and diversity analyses to 16S amplicon data using QIIME. We recommend first working through the <a class="reference external" href="./illumina_overview_tutorial.html">QIIME Illumina Overview Tutorial</a>, which covers many of these analysis steps at a higher level. This tutorial provides additional detail on the steps that are being performed by the QIIME workflows.</p>
<p>As an example, we will use data from a study of the response of mouse gut microbial communities to fasting (<a class="reference external" href="http://www.ncbi.nlm.nih.gov/pubmed/19549860">Crawford et al. (2009)</a>). To make this tutorial run quickly on a personal computer, we will use a subset of the data generated from 5 animals kept on the control <em>ad libitum</em> fed diet, and 4 animals fasted for 24 hours before sacrifice. At the end of our tutorial, we will be able to compare the community structure of control vs. fasted animals. In particular, we will be able to compare taxonomic profiles for each sample type and perform alpha and beta diversity analyses to look for community-level differences in the samples.</p>
<p>If you need help while working through the tutorial, please visit <a class="reference external" href="http://help.qiime.org">http://help.qiime.org</a>.</p>
<div class="section" id="obtaining-the-data">
<h2>Obtaining the data<a class="headerlink" href="#obtaining-the-data" title="Permalink to this headline">¶</a></h2>
<p>All the files you will need for this tutorial can be downloaded <a class="reference external" href="ftp://ftp.microbio.me/pub/qiime-files/qiime_overview_tutorial.zip">here</a>. Descriptions of these files are below.</p>
<p>Open a terminal and use the <tt class="docutils literal"><span class="pre">cd</span></tt> command to move to the directory where you downloaded the tutorial data. Next, unzip the tutorial data and move into the unzipped directory that is created:</p>
<div class="highlight-python"><div class="highlight"><pre>unzip qiime_overview_tutorial.zip
cd qiime_overview_tutorial
</pre></div>
</div>
<div class="section" id="sequences-fna">
<h3>Sequences (.fna)<a class="headerlink" href="#sequences-fna" title="Permalink to this headline">¶</a></h3>
<p>This is the 454-machine generated FASTA file. Using the Amplicon processing software on the 454 FLX standard, each region of the PTP plate will yield a fasta file of form <tt class="file docutils literal"><span class="pre">1.TCA.454Reads.fna</span></tt>, where &#8220;1&#8221; is replaced with the appropriate region number. For the purposes of this tutorial, we will use the fasta file <tt class="file docutils literal"><span class="pre">Fasting_Example.fna</span></tt>.</p>
</div>
<div class="section" id="quality-scores-qual">
<h3>Quality Scores (.qual)<a class="headerlink" href="#quality-scores-qual" title="Permalink to this headline">¶</a></h3>
<p>This is the 454-machine generated quality score file, which contains a score for each base in each sequence included in the FASTA file. Like the fasta file mentioned above, the Amplicon processing software will generate one of these files for each region of the PTP plate, named <tt class="file docutils literal"><span class="pre">1.TCA.454Reads.qual</span></tt>, etc. For the purposes of this tutorial, we will use the quality scores file <tt class="file docutils literal"><span class="pre">Fasting_Example.qual</span></tt>.</p>
</div>
<div class="section" id="mapping-file-tab-delimited-txt">
<h3>Mapping File (Tab-delimited .txt)<a class="headerlink" href="#mapping-file-tab-delimited-txt" title="Permalink to this headline">¶</a></h3>
<p>The mapping file is generated by the user. This file contains all of the information about the samples necessary to perform the data analysis. See <a class="reference external" href="http://qiime.org/documentation/file_formats.html#metadata-mapping-files">here</a> for more details about the mapping file format.</p>
<p>For the purposes of this tutorial, we will use the mapping file <tt class="file docutils literal"><span class="pre">Fasting_Map.txt</span></tt>. The contents of the mapping file are shown here - as you can see, a nucleotide barcode sequence is provided for each of the 9 samples, as well as metadata related to treatment group and date of birth, and general run descriptions about the project:</p>
<div class="highlight-python"><div class="highlight"><pre>#SampleID  BarcodeSequence LinkerPrimerSequence    Treatment DOB   Description
#Example mapping file for the QIIME analysis package. These 9 samples are from a study of the effects of
#exercise and diet on mouse cardiac physiology (Crawford, et al, PNAS, 2009).
PC.354 AGCACGAGCCTA    YATGCTGCCTCCCGTAGGAGT   Control 20061218    Control_mouse__I.D._354
PC.355 AACTCGTCGATG    YATGCTGCCTCCCGTAGGAGT   Control 20061218    Control_mouse__I.D._355
PC.356 ACAGACCACTCA    YATGCTGCCTCCCGTAGGAGT   Control 20061126    Control_mouse__I.D._356
PC.481 ACCAGCGACTAG    YATGCTGCCTCCCGTAGGAGT   Control 20070314    Control_mouse__I.D._481
PC.593 AGCAGCACTTGT    YATGCTGCCTCCCGTAGGAGT   Control 20071210    Control_mouse__I.D._593
PC.607 AACTGTGCGTAC    YATGCTGCCTCCCGTAGGAGT   Fast    20071112    Fasting_mouse__I.D._607
PC.634 ACAGAGTCGGCT    YATGCTGCCTCCCGTAGGAGT   Fast    20080116    Fasting_mouse__I.D._634
PC.635 ACCGCAGAGTCA    YATGCTGCCTCCCGTAGGAGT   Fast    20080116    Fasting_mouse__I.D._635
PC.636 ACGGTGAGTGTC    YATGCTGCCTCCCGTAGGAGT   Fast    20080116    Fasting_mouse__I.D._636
</pre></div>
</div>
</div>
</div>
<div class="section" id="validate-the-mapping-file">
<span id="checkmapping"></span><h2>Validate the mapping file<a class="headerlink" href="#validate-the-mapping-file" title="Permalink to this headline">¶</a></h2>
<p>First, you should ensure that your mapping file is formatted correctly with the <a class="reference external" href="../scripts/validate_mapping_file.html">validate_mapping_file.py</a> script:</p>
<div class="highlight-python"><div class="highlight"><pre>validate_mapping_file.py -m Fasting_Map.txt -o mapping_output
</pre></div>
</div>
<p>This script will print a message indicating whether or not problems were found in the mapping file. An HTML file showing the location of errors and warnings will be generated in the output directory, and a plain text log file will also be created. Errors will cause fatal problems with subsequent scripts and must be corrected before moving forward. Warnings will not cause fatal problems, but it is encouraged that you fix these problems as they are often indicative of typos in your mapping file, invalid characters, or other unintended errors that will impact downstream analysis. A file ending with <tt class="file docutils literal"><span class="pre">_corrected.txt</span></tt> will also be created in the output directory, which will have a copy of the mapping file with invalid characters replaced by underscores.</p>
<p>Reverse primers can be specified in the mapping file, for removal during the demultiplexing step.  This is not required, but it is <strong>strongly</strong> recommended, as leaving in sequences following primers, such as sequencing adapters, can interfere with OTU picking and taxonomic assignment.</p>
<p>For example, a mapping file with reverse primers specified in the <tt class="docutils literal"><span class="pre">ReversePrimer</span></tt> column might look like:</p>
<div class="highlight-python"><div class="highlight"><pre>#SampleID  BarcodeSequence LinkerPrimerSequence    Treatment ReversePrimer   Description
#Example mapping file for the QIIME analysis package. These 9 samples are from a study of the effects of
#exercise and diet on mouse cardiac physiology (Crawford, et al, PNAS, 2009).
PC.354 AGCACGAGCCTA    YATGCTGCCTCCCGTAGGAGT   Control GCGCACGGGTGAGTA    Control_mouse__I.D._354
PC.355 AACTCGTCGATG    YATGCTGCCTCCCGTAGGAGT   Control GCGCACGGGTGAGTA    Control_mouse__I.D._355
PC.356 ACAGACCACTCA    YATGCTGCCTCCCGTAGGAGT   Control GCGCACGGGTGAGTA    Control_mouse__I.D._356
PC.481 ACCAGCGACTAG    YATGCTGCCTCCCGTAGGAGT   Control GCGCACGGGTGAGTA    Control_mouse__I.D._481
PC.593 AGCAGCACTTGT    YATGCTGCCTCCCGTAGGAGT   Control GCGCACGGGTGAGTA    Control_mouse__I.D._593
PC.607 AACTGTGCGTAC    YATGCTGCCTCCCGTAGGAGT   Fast    GCGCACGGGTGAGTA    Fasting_mouse__I.D._607
PC.634 ACAGAGTCGGCT    YATGCTGCCTCCCGTAGGAGT   Fast    GCGCACGGGTGAGTA    Fasting_mouse__I.D._634
PC.635 ACCGCAGAGTCA    YATGCTGCCTCCCGTAGGAGT   Fast    GCGCACGGGTGAGTA    Fasting_mouse__I.D._635
PC.636 ACGGTGAGTGTC    YATGCTGCCTCCCGTAGGAGT   Fast    GCGCACGGGTGAGTA    Fasting_mouse__I.D._636
</pre></div>
</div>
<p>The reverse primers, like the forward primers, are written in 5&#8217;-&gt;3&#8217; direction.  In this case, these are not the true reverse primers used, but rather just a somewhat conserved site in the sequences used for this example.</p>
<p>The primer construct and amplicon used in <a class="reference external" href="http://www.ncbi.nlm.nih.gov/pubmed/19549860">Crawford et al. (2009)</a> is shown below, annotated using QIIME nomenclature:</p>
<img alt="../_images/example_primer_construct.png" class="align-center" src="../_images/example_primer_construct.png" />
<p>454 sequencing, in most cases, generates sequences that begin at the <tt class="docutils literal"><span class="pre">BarcodeSequence</span></tt>, which is followed by the <tt class="docutils literal"><span class="pre">LinkerPrimerSequence</span></tt>, both of which are automatically removed during the demultiplexing step described below.  The <tt class="docutils literal"><span class="pre">ReversePrimer</span></tt> (i.e., the primer at the end of the read) is not removed by default but can be using the <tt class="docutils literal"><span class="pre">-z</span></tt> option to <a class="reference external" href="../scripts/split_libraries.html">split_libraries.py</a>. The adapter sequence (Adapter B) is not a biological sequence, so must be removed if you suspect that it is contained within your reads. This can be achieved using the <tt class="docutils literal"><span class="pre">-z</span></tt> option to <a class="reference external" href="../scripts/split_libraries.html">split_libraries.py</a>.</p>
</div>
<div class="section" id="demultiplex-and-quality-filter-reads">
<span id="assignsamples"></span><h2>Demultiplex and quality filter reads<a class="headerlink" href="#demultiplex-and-quality-filter-reads" title="Permalink to this headline">¶</a></h2>
<p>The next task is to assign the multiplexed reads to samples based on their nucleotide barcode (this is known as <em>demultiplexing</em>). This step also performs quality filtering based on the characteristics of each sequence, removing any low quality or ambiguous reads. To perform these steps we&#8217;ll use <a class="reference external" href="../scripts/split_libraries.html">split_libraries.py</a>:</p>
<div class="highlight-python"><div class="highlight"><pre>split_libraries.py -m Fasting_Map.txt -f Fasting_Example.fna -q Fasting_Example.qual -o split_library_output
</pre></div>
</div>
<p>This will create three files in the new directory <tt class="file docutils literal"><span class="pre">split_library_output/</span></tt>:</p>
<ul class="simple">
<li><tt class="file docutils literal"><span class="pre">split_library_log.txt</span></tt> : This file contains the summary of demultiplexing and quality filtering, including the number of reads detected for each sample and a brief summary of any reads that were removed due to quality considerations.</li>
<li><tt class="file docutils literal"><span class="pre">histograms.txt</span></tt> : This tab-delimited file shows the number of reads at regular size intervals before and after splitting the library.</li>
<li><tt class="file docutils literal"><span class="pre">seqs.fna</span></tt> : This is a fasta formatted file where each sequence is renamed according to the sample it came from. The header line also contains the name of the read in the input fasta file and information on any barcode errors that were corrected.</li>
</ul>
<p>A few lines from the <tt class="file docutils literal"><span class="pre">seqs.fna</span></tt> file are shown below:</p>
<div class="highlight-python"><div class="highlight"><pre>&gt;PC.634_1 FLP3FBN01ELBSX orig_bc=ACAGAGTCGGCT new_bc=ACAGAGTCGGCT bc_diffs=0
CTGGGCCGTGTCTCAGTCCCAATGTGGCCGTTTACCCTCTCAGGCCGGCTACGCATCATCGCC....
&gt;PC.634_2 FLP3FBN01EG8AX orig_bc=ACAGAGTCGGCT new_bc=ACAGAGTCGGCT bc_diffs=0
TTGGACCGTGTCTCAGTTCCAATGTGGGGGCCTTCCTCTCAGAACCCCTATCCATCGAAGGCTT....
&gt;PC.354_3 FLP3FBN01EEWKD orig_bc=AGCACGAGCCTA new_bc=AGCACGAGCCTA bc_diffs=0
TTGGGCCGTGTCTCAGTCCCAATGTGGCCGATCAGTCTCTTAACTCGGCTATGCATCATTGCCTT....
&gt;PC.481_4 FLP3FBN01DEHK3 orig_bc=ACCAGCGACTAG new_bc=ACCAGCGACTAG bc_diffs=0
CTGGGCCGTGTCTCAGTCCCAATGTGGCCGTTCAACCTCTCAGTCCGGCTACTGATCGTCGACT....
</pre></div>
</div>
<p>The first several lines of the <tt class="file docutils literal"><span class="pre">split_library_log.txt</span></tt> file are shown below:</p>
<div class="highlight-python"><div class="highlight"><pre>Number raw input seqs   1339

Length outside bounds of 200 and 1000   0
Num ambiguous bases exceeds limit of 6  0
Missing Qual Score      0
Mean qual score below minimum of 25     1
Max homopolymer run exceeds limit of 6  0
Num mismatches in primer exceeds limit of 0: 1
...
</pre></div>
</div>
<p>Reverse primer removal can be accomplished by adding the <tt class="docutils literal"><span class="pre">-z</span></tt> option.  An example command using the mapping file with reverse primers described above:</p>
<div class="highlight-python"><div class="highlight"><pre>split_libraries.py -m Fasting_Map_reverse_primers.txt -f Fasting_Example.fna -q Fasting_Example.qual -z truncate_only -o split_library_output_revprimers
</pre></div>
</div>
<p>If the number of sequences where the reverse primer is not identifiable is high, you should check the primer sequence to make sure it is in 5&#8217;-&gt;3&#8217; orientation, or increase the number of mismatches allowed with <tt class="docutils literal"><span class="pre">--reverse_primer_mismatches</span></tt>.</p>
<p>Data that are already demultiplexed can have reverse primers removed using the standalone script <a class="reference external" href="../scripts/truncate_reverse_primer.html">truncate_reverse_primer.py</a>.</p>
</div>
<div class="section" id="de-novo-otu-picking">
<span id="pickotusandrepseqs"></span><h2>De novo OTU picking<a class="headerlink" href="#de-novo-otu-picking" title="Permalink to this headline">¶</a></h2>
<p>Here we will be running the <a class="reference external" href="../scripts/pick_de_novo_otus.html">pick_de_novo_otus.py</a> workflow, which runs a series of other scripts automatically. This workflow consists of the following steps:</p>
<ol class="arabic simple">
<li>Pick OTUs based on sequence similarity within the reads (<a class="reference external" href="../scripts/pick_otus.html">pick_otus.py</a>)</li>
<li>Pick a representative sequence for each OTU (<a class="reference external" href="../scripts/pick_rep_set.html">pick_rep_set.py</a>)</li>
<li>Assign taxonomy to OTU representative sequences (<a class="reference external" href="../scripts/assign_taxonomy.html">assign_taxonomy.py</a>)</li>
<li>Align OTU representative sequences (<a class="reference external" href="../scripts/align_seqs.html">align_seqs.py</a>)</li>
<li>Filter the alignment (<a class="reference external" href="../scripts/filter_alignment.html">filter_alignment.py</a>)</li>
<li>Build a phylogenetic tree  (<a class="reference external" href="../scripts/make_phylogeny.html">make_phylogeny.py</a>)</li>
<li>Make the OTU table (<a class="reference external" href="../scripts/make_otu_table.html">make_otu_table.py</a>)</li>
</ol>
<p>Using the output from <a class="reference external" href="../scripts/split_libraries.html">split_libraries.py</a> (<tt class="file docutils literal"><span class="pre">seqs.fna</span></tt>), run the following command:</p>
<div class="highlight-python"><div class="highlight"><pre>pick_de_novo_otus.py -i split_library_output/seqs.fna -o otus
</pre></div>
</div>
<p>If you&#8217;re interested in denoising 454 data, refer to the <a class="reference external" href="./denoising_454_data.html">denoising tutorial</a>.</p>
<p>The results of <a class="reference external" href="../scripts/pick_de_novo_otus.html">pick_de_novo_otus.py</a> are in <tt class="file docutils literal"><span class="pre">otus/</span></tt>, and a description of the steps performed and the results follow:</p>
<div class="section" id="step-1-pick-otus-based-on-sequence-similarity-within-the-reads">
<span id="pickotusseqsim"></span><h3>Step 1. Pick OTUs based on sequence similarity within the reads<a class="headerlink" href="#step-1-pick-otus-based-on-sequence-similarity-within-the-reads" title="Permalink to this headline">¶</a></h3>
<p>At this step, all of the sequences from all of the samples will be clustered into Operational Taxonomic Units (OTUs) based on their sequence similarity. OTUs in QIIME are clusters of sequences, frequently intended to represent some degree of taxonomic relatedness. For example, when sequences are clustered at 97% sequence similarity, each resulting cluster is typically thought of as representing a species. This model and the current techniques for picking OTUs are known to be flawed, however, in that 97% OTUs do not match what humans have called species for many microbes. Determining exactly how OTUs should be defined, and what they represent, is an active area of research. See <a class="reference external" href="otu_picking.html">here</a> for more information about OTU picking with QIIME.</p>
</div>
<div class="section" id="step-2-pick-a-representative-sequence-for-each-otu">
<span id="pickrepseqsforotu"></span><h3>Step 2. Pick a representative sequence for each OTU<a class="headerlink" href="#step-2-pick-a-representative-sequence-for-each-otu" title="Permalink to this headline">¶</a></h3>
<p>Since each OTU may be made up of many related sequences, we will pick a representative sequence from each OTU for downstream analysis. This representative sequence will be used for taxonomic identification of the OTU and phylogenetic alignment. QIIME uses the OTU file created above and extracts a representative sequence from the fasta file by one of several methods.</p>
<p>In the <tt class="file docutils literal"><span class="pre">otus/rep_set/</span></tt> directory, QIIME has a fasta file <tt class="file docutils literal"><span class="pre">seqs_rep_set.fasta</span></tt> containing one representative sequence for each OTU. In this fasta file, the sequence has been renamed with the OTU identifier, and the additional information on the header line reflects the sequence used as the representative:</p>
<div class="highlight-python"><div class="highlight"><pre>&gt;0 PC.636_424
CTGGGCCGTATCTCAGTCCCAATGTGGCCGGTCGACCTCTC....
&gt;1 PC.481_321
TTGGGCCGTGTCTCAGTCCCAATGTGGCCGTCCGCCCTCTC....
</pre></div>
</div>
</div>
<div class="section" id="step-3-assign-taxonomy-to-otu-representative-sequences">
<span id="assigntax"></span><h3>Step 3. Assign taxonomy to OTU representative sequences<a class="headerlink" href="#step-3-assign-taxonomy-to-otu-representative-sequences" title="Permalink to this headline">¶</a></h3>
<p>Next, taxonomy will be assigned to each representative sequence. By default, QIIME uses the uclust consensus taxonomy classifier to attempt to assign taxonomy to each representative sequence resulting from step 2.</p>
<p>In the directory <tt class="file docutils literal"><span class="pre">otus/uclust_assigned_taxonomy/</span></tt>, there will be a log file and a text file. The text file (which we refer to as an observation or OTU metadata file) contains a line for each OTU, followed by the taxonomic assignment, the fraction of uclust hits that contained this taxonomic assignment, and the number of uclust hits that were found. For some OTUs, the assignment may be as specific as a bacterial species, while others may not be assignable at all (and will therefore be labeled as <em>Unassigned</em>). Below are the first few lines of an observation metadata file containing the results of uclust taxonomic assignment:</p>
<div class="highlight-python"><div class="highlight"><pre>denovo367       k__Bacteria; p__Bacteroidetes; c__Bacteroidia; o__Bacteroidales; f__S24-7; g__; s__     1.00    3
denovo366       k__Bacteria; p__Firmicutes; c__Clostridia; o__Clostridiales; f__; g__; s__      1.00    3
denovo365       k__Bacteria; p__Firmicutes; c__Erysipelotrichi; o__Erysipelotrichales; f__Erysipelotrichaceae; g__Allobaculum; s__      1.00    3
denovo364       k__Bacteria; p__Firmicutes; c__Clostridia; o__Clostridiales; f__Lachnospiraceae 0.67    3
denovo281       Unassigned      1.00    1
</pre></div>
</div>
</div>
<div class="section" id="step-4-align-otu-representative-sequences">
<span id="alignotuseq"></span><h3>Step 4. Align OTU representative sequences<a class="headerlink" href="#step-4-align-otu-representative-sequences" title="Permalink to this headline">¶</a></h3>
<p>Alignment of the OTU representative sequences and phylogeny inference is necessary only if phylogenetic metrics such as <a class="reference external" href="http://www.ncbi.nlm.nih.gov/pubmed/16332807">UniFrac</a> will be subsequently invoked. Alignments can either be generated de novo using programs such as MUSCLE, or through alignment to an existing alignment with tools like <a class="reference external" href="http://biocore.github.io/pynast/">PyNAST</a>. For small studies such as this tutorial, either method is possible. However, for studies involving many sequences (roughly, more than 1000), de novo aligners are very slow and alignment with <a class="reference external" href="http://biocore.github.io/pynast/">PyNAST</a> is necessary. Since alignment is one of the most computationally intensive bottlenecks in the pipeline, large studies benefit greatly from parallelization of this task, which is possible with <a class="reference external" href="http://biocore.github.io/pynast/">PyNAST</a>.</p>
<p>After aligning the sequences, a log file and an alignment file are created in the directory <tt class="file docutils literal"><span class="pre">otus/pynast_aligned_seqs/</span></tt>.</p>
</div>
<div class="section" id="step-5-filter-the-alignment">
<span id="filteraln"></span><h3>Step 5. Filter the alignment<a class="headerlink" href="#step-5-filter-the-alignment" title="Permalink to this headline">¶</a></h3>
<p>Before inferring a phylogenetic tree relating the sequences, it is beneficial to filter the sequence alignment to remove columns comprised of only gaps, and locations known to be excessively variable. QIIME uses a 16S alignment Lane mask (Lane, D.J. 1991) by default. After filtering, a filtered alignment file is created in the directory <tt class="file docutils literal"><span class="pre">otus/pynast_aligned_seqs/</span></tt>.</p>
</div>
<div class="section" id="step-6-build-a-phylogenetic-tree">
<span id="maketree"></span><h3>Step 6. Build a phylogenetic tree<a class="headerlink" href="#step-6-build-a-phylogenetic-tree" title="Permalink to this headline">¶</a></h3>
<p>The filtered alignment file produced in the directory <tt class="file docutils literal"><span class="pre">otus/pynast_aligned_seqs/</span></tt> is then used to build a phylogenetic tree using a tree-building program.</p>
<p>The Newick format tree file is written to <tt class="file docutils literal"><span class="pre">rep_set.tre</span></tt>, which is located in the <tt class="file docutils literal"><span class="pre">otus/</span></tt> directory . This file can be viewed in tree visualization software, and is necessary for <a class="reference external" href="http://www.ncbi.nlm.nih.gov/pubmed/16332807">UniFrac</a> diversity measurements and other phylogenetically-aware analyses (described below). The tree obtained can be visualized with programs such as <a class="reference external" href="http://tree.bio.ed.ac.uk/software/figtree/">FigTree</a>, which was used to visualize the phylogenetic tree stored in <tt class="file docutils literal"><span class="pre">rep_set.tre</span></tt>:</p>
<img alt="../_images/tree.png" class="align-center" src="../_images/tree.png" />
</div>
<div class="section" id="step-7-make-the-otu-table">
<span id="makeotutable"></span><h3>Step 7. Make the OTU table<a class="headerlink" href="#step-7-make-the-otu-table" title="Permalink to this headline">¶</a></h3>
<p>Using taxonomic assignments (step 3) and the OTU map (step 1) QIIME assembles a table of OTU abundances in each sample with taxonomic identifiers for each OTU.</p>
<p>The result of this step is <tt class="file docutils literal"><span class="pre">otu_table.biom</span></tt>, which is located in the <tt class="file docutils literal"><span class="pre">otus/</span></tt> directory. For more information about the OTU table format, which is stored in the BIOM format, please see <a class="reference external" href="http://biom-format.org">here</a>.</p>
</div>
</div>
<div class="section" id="summarize-the-otu-table">
<span id="perlibrarystats"></span><h2>Summarize the OTU table<a class="headerlink" href="#summarize-the-otu-table" title="Permalink to this headline">¶</a></h2>
<p>To view summary statistics of the OTU table, run:</p>
<div class="highlight-python"><div class="highlight"><pre>biom summarize-table -i otus/otu_table.biom
</pre></div>
</div>
<p>The summary shows that there are relatively few sequences in this tutorial example, but the sequences present are fairly evenly distributed among the 9 microbial communities:</p>
<div class="highlight-python"><div class="highlight"><pre>Num samples: 9
Num observations: 419
Total count: 1337
Table density (fraction of non-zero values): 0.168

Counts/sample summary:
 Min: 146.0
 Max: 150.0
 Median: 149.000
 Mean: 148.556
 Std. dev.: 1.257
 Sample Metadata Categories:
 Observation Metadata Categories: taxonomy

Counts/sample detail:
 PC.481: 146.0
 PC.355: 147.0
 PC.636: 148.0
 PC.607: 149.0
 PC.635: 149.0
 PC.593: 149.0
 PC.354: 149.0
 PC.634: 150.0
 PC.356: 150.0
</pre></div>
</div>
</div>
<div class="section" id="make-an-otu-network">
<span id="makeotunetwork"></span><h2>Make an OTU network<a class="headerlink" href="#make-an-otu-network" title="Permalink to this headline">¶</a></h2>
<p>To create an OTU network, using the following command:</p>
<div class="highlight-python"><div class="highlight"><pre>make_otu_network.py -m Fasting_Map.txt -i otus/otu_table.biom -o otus
</pre></div>
</div>
<p>To visualize the network, we use the <a class="reference external" href="http://www.cytoscape.org/">Cytoscape</a> program (which you can run by calling cytoscape from the command line &#8211; you may need to call this beginning either with a capital or lowercase &#8216;C&#8217; depending on your version of <a class="reference external" href="http://www.cytoscape.org/">Cytoscape</a>), where each red circle represents a sample and each white square represents an OTU. The lines represent the OTUs present in a particular sample (blue for controls and green for fasting). For more information about opening the files in <a class="reference external" href="http://www.cytoscape.org/">Cytoscape</a> please refer to <a class="reference external" href="./making_cytoscape_networks.html">Making Cytoscape Networks</a>.</p>
<img alt="../_images/network.png" class="align-center" src="../_images/network.png" />
</div>
<div class="section" id="summarize-communities-by-taxonomic-composition">
<span id="summarizetaxa"></span><h2>Summarize communities by taxonomic composition<a class="headerlink" href="#summarize-communities-by-taxonomic-composition" title="Permalink to this headline">¶</a></h2>
<p>You can group OTUs by different taxonomic levels (phylum, class, family, etc.) with the workflow script <a class="reference external" href="../scripts/summarize_taxa_through_plots.html">summarize_taxa_through_plots.py</a>. Note that this process depends directly on the method used to assign taxonomic information to OTUS (see <a class="reference internal" href="#assigntax">Assigning Taxonomy</a> above):</p>
<div class="highlight-python"><div class="highlight"><pre>summarize_taxa_through_plots.py -i otus/otu_table.biom -o taxa_summary -m Fasting_Map.txt
</pre></div>
</div>
<p>The script will generate new tables at various taxonomic levels (we&#8217;ll refer to these as <em>taxonomy tables</em>, which are different than <em>OTU tables</em>). For example, the class-level table is located at <tt class="file docutils literal"><span class="pre">taxa_summary/otu_table_L3.txt</span></tt>. Each taxonomy table contains the relative abundances of taxa within each sample:</p>
<div class="highlight-python"><div class="highlight"><pre>#OTU ID     PC.636  PC.635 PC.356   PC.481  PC.354 PC.593   PC.355  PC.607 PC.634
Unassigned;Other;Other 0.027027027027 0.00671140939597 0.0133333333333 0.00684931506849     0.0 0.00671140939597 0.00680272108844 0.0134228187919   0.02
k__Bacteria;Other;Other     0.0 0.0 0.0     0.0     0.0     0.0     0.0 0.00671140939597    0.0
k__Bacteria;p__Actinobacteria;c__Coriobacteriia 0.00675675675676    0.0     0.0 0.00684931506849    0.0     0.0     0.0 0.0134228187919 0.0133333333333
k__Bacteria;p__Bacteroidetes;c__Bacteroidia 0.675675675676 0.530201342282   0.2 0.143835616438 0.0805369127517 0.389261744966 0.285714285714 0.288590604027 0.64
k__Bacteria;p__Deferribacteres;c__Deferribacteres 0.0472972972973 0.0134228187919   0.0     0.0     0.0 0.0 0.0     0.0201342281879 0.0333333333333
k__Bacteria;p__Firmicutes;c__Bacilli        0.027027027027  0.0 0.0933333333333 0.089041095890 0.107382550336 0.0335570469799 0.0136054421769 0.00671140939597      0.02
k__Bacteria;p__Firmicutes;c__Clostridia     0.195945945946 0.436241610738 0.686666666667 0.712328767123 0.798657718121 0.389261744966 0.69387755102 0.469798657718 0.213333333333
k__Bacteria;p__Firmicutes;c__Erysipelotrichi 0.0135135135135 0.00671140939597       0.0 0.0342465753425 0.0134228187919 0.147651006711      0.0 0.154362416107 0.0266666666667
</pre></div>
</div>
<p id="maketaxacharts">To view the resulting charts, open the area or bar chart html file located in the <tt class="file docutils literal"><span class="pre">taxa_summary/taxa_summary_plots</span></tt> folder. The following chart shows the taxonomic assignments for each sample as a bar chart. You can mouse-over the plot to see which taxa are contributing to the percentage shown:</p>
<img alt="../_images/barchart1.png" class="align-center" src="../_images/barchart1.png" />
</div>
<div class="section" id="make-a-taxonomy-heatmap">
<span id="makeheatmap"></span><h2>Make a taxonomy heatmap<a class="headerlink" href="#make-a-taxonomy-heatmap" title="Permalink to this headline">¶</a></h2>
<p>QIIME supports generating heatmap images of BIOM tables (e.g., OTU tables or the taxonomy tables generated in the previous step) with <a class="reference external" href="../scripts/make_otu_heatmap.html">make_otu_heatmap.py</a>. Let&#8217;s create a heatmap illustrating class-level abundances on a per-sample basis, where samples are sorted by whether they are from control or fasted mice:</p>
<div class="highlight-python"><div class="highlight"><pre>make_otu_heatmap.py -i taxa_summary/otu_table_L3.biom -o taxa_summary/otu_table_L3_heatmap.pdf -c Treatment -m Fasting_Map.txt
</pre></div>
</div>
<p>A PDF file is created as <tt class="file docutils literal"><span class="pre">taxa_summary/otu_table_L3_heatmap.pdf</span></tt>. The first four samples are from fasted mice and the last five are from controls. This clearly illustrates class-level differences in the taxonomic composition of the samples:</p>
<img alt="../_images/heatmap.png" class="align-center" src="../_images/heatmap.png" />
</div>
<div class="section" id="compute-alpha-diversity-and-generate-alpha-rarefaction-plots">
<span id="compalphadivrarecurves"></span><h2>Compute alpha diversity and generate alpha rarefaction plots<a class="headerlink" href="#compute-alpha-diversity-and-generate-alpha-rarefaction-plots" title="Permalink to this headline">¶</a></h2>
<p>Community ecologists are often interested in computing <em>alpha</em> (or the <em>within-sample</em>) diversity for samples or groups of samples in their study. Here, we will determine the level of alpha diversity in our samples using QIIME&#8217;s <a class="reference external" href="../scripts/alpha_rarefaction.html">alpha_rarefaction.py</a> workflow, which performs the following steps:</p>
<ol class="arabic simple">
<li>Generate rarefied OTU tables (<a class="reference external" href="../scripts/multiple_rarefactions.html">multiple_rarefactions.py</a>)</li>
<li>Compute measures of alpha diversity for each rarefied OTU table (<a class="reference external" href="../scripts/alpha_diversity.html">alpha_diversity.py</a>)</li>
<li>Collate alpha diversity results (<a class="reference external" href="../scripts/collate_alpha.html">collate_alpha.py</a>)</li>
<li>Generate alpha rarefaction plots (<a class="reference external" href="../scripts/make_rarefaction_plots.html">make_rarefaction_plots.py</a>)</li>
</ol>
<p>Although we could run this workflow with the (sensible) default parameters, this provides an opportunity to illustrate the use of custom parameters in a QIIME workflow. To see what measures of alpha diversity will be computed by default, run:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">alpha_diversity</span><span class="o">.</span><span class="n">py</span> <span class="o">-</span><span class="n">h</span>
</pre></div>
</div>
<p>You should see, among other information:</p>
<div class="highlight-python"><div class="highlight"><pre>-m METRICS, --metrics=METRICS
    Alpha-diversity metric(s) to use. A comma-separated
    list should be provided when multiple metrics are
    specified. [default: PD_whole_tree,chao1,observed_otus]
</pre></div>
</div>
<p>which indicates that the metrics that will be used by default are <tt class="docutils literal"><span class="pre">PD_whole_tree</span></tt>, <tt class="docutils literal"><span class="pre">chao1</span></tt>, and <tt class="docutils literal"><span class="pre">observed_otus</span></tt>. If we additionally wanted to compute Shannon Index, we could create a parameters file (which for the sake of this example we&#8217;ll call <tt class="file docutils literal"><span class="pre">alpha_params.txt</span></tt>) containing the following line:</p>
<div class="highlight-python"><div class="highlight"><pre>alpha_diversity:metrics shannon,PD_whole_tree,chao1,observed_otus
</pre></div>
</div>
<p>For more information on creating parameters files, see <a class="reference external" href="../documentation/qiime_parameters_files.html">here</a>.</p>
<p>We can next run <a class="reference external" href="../scripts/alpha_rarefaction.html">alpha_rarefaction.py</a>, which requires the OTU table (<tt class="docutils literal"><span class="pre">-i</span></tt>) and phylogenetic tree (<tt class="docutils literal"><span class="pre">-t</span></tt>) from <a class="reference internal" href="#pickotusandrepseqs">above</a>, and the parameters file we just created:</p>
<div class="highlight-python"><div class="highlight"><pre>alpha_rarefaction.py -i otus/otu_table.biom -m Fasting_Map.txt -o arare -p alpha_params.txt -t otus/rep_set.tre
</pre></div>
</div>
<p>Descriptions of the steps involved in <a class="reference external" href="../scripts/alpha_rarefaction.html">alpha_rarefaction.py</a> follow:</p>
<div class="section" id="step-1-generate-rarefied-otu-tables">
<span id="rareotutable"></span><h3>Step 1. Generate rarefied OTU tables<a class="headerlink" href="#step-1-generate-rarefied-otu-tables" title="Permalink to this headline">¶</a></h3>
<p>The directory <tt class="file docutils literal"><span class="pre">arare/rarefaction/</span></tt> will contain many text files named <tt class="file docutils literal"><span class="pre">rarefaction_##_#.txt</span></tt>; the first set of numbers represents the number of sequences sampled, and the last number represents the iteration number. If you opened one of these files, you would find an OTU table where for each sample the sum of the counts equals the number of samples taken.</p>
<p>To keep the results of <a class="reference external" href="../scripts/alpha_rarefaction.html">alpha_rarefaction.py</a> to a manageable size, these results are deleted unless you pass the <tt class="docutils literal"><span class="pre">--retain_intermediate_files</span></tt> option to <a class="reference external" href="../scripts/alpha_rarefaction.html">alpha_rarefaction.py</a>.</p>
</div>
<div class="section" id="step-2-compute-measures-of-alpha-diversity-for-each-rarefied-otu-table">
<span id="computealphadiv"></span><h3>Step 2. Compute measures of alpha diversity for each rarefied OTU table<a class="headerlink" href="#step-2-compute-measures-of-alpha-diversity-for-each-rarefied-otu-table" title="Permalink to this headline">¶</a></h3>
<p>The rarefied tables are the basis for calculating alpha diversity metrics, which describe the richness and/or evenness of taxa in a single sample. QIIME allows users to calculate more than two dozen different diversity metrics. The full list of available metrics is available <a class="reference external" href="http://scikit-bio.org/docs/latest/generated/skbio.diversity.alpha.html">here</a>. Each metric has different strengths and limitations. Technical discussion of each metric is readily available online and in ecology textbooks, but it is beyond the scope of this tutorial. By default, QIIME calculates three metrics: Chao1 (<tt class="docutils literal"><span class="pre">chao1</span></tt>), Observed OTUs (<tt class="docutils literal"><span class="pre">observed_otus</span></tt>, previously known as Observed Species), and Phylogenetic Diversity (<tt class="docutils literal"><span class="pre">PD_whole_tree</span></tt>). In addition, in the <tt class="file docutils literal"><span class="pre">alpha_params.txt</span></tt> file we added the Shannon Index (<tt class="docutils literal"><span class="pre">shannon</span></tt>) to the list of alpha diversity measures that we calculated here.</p>
<p>The result of this step produces text files with the results of the alpha diversity computations performed on the rarefied OTU tables. The results are located in the <tt class="file docutils literal"><span class="pre">arare/alpha_div/</span></tt> directory.</p>
<p>To keep the results of <a class="reference external" href="../scripts/alpha_rarefaction.html">alpha_rarefaction.py</a> to a manageable size, these results are deleted unless you pass the <tt class="docutils literal"><span class="pre">--retain_intermediate_files</span></tt> option to <a class="reference external" href="../scripts/alpha_rarefaction.html">alpha_rarefaction.py</a>.</p>
</div>
<div class="section" id="step-3-collate-alpha-diversity-results">
<span id="collateotutable"></span><h3>Step 3. Collate alpha diversity results<a class="headerlink" href="#step-3-collate-alpha-diversity-results" title="Permalink to this headline">¶</a></h3>
<p>The output directory <tt class="file docutils literal"><span class="pre">arare/alpha_div/</span></tt> will contain one text file <tt class="file docutils literal"><span class="pre">alpha_rarefaction_##_#</span></tt> for every file input from <tt class="file docutils literal"><span class="pre">arare/rarefaction/</span></tt>, where the numbers represent the number of samples and iterations as before. The content of this tab delimited file is the calculated metrics for each sample. To collapse the individual files into a single combined table, the workflow uses <a class="reference external" href="../scripts/collate_alpha.html">collate_alpha.py</a>.</p>
<p>In the output directory <tt class="file docutils literal"><span class="pre">arare/alpha_div_collated/</span></tt>, there will be one file for every alpha diversity metric used. Each file will contain the alpha diversity measure for every sample, arranged in ascending order from lowest number of sequences per sample to highest. A portion of the <tt class="file docutils literal"><span class="pre">observed_otus.txt</span></tt> file are shown below:</p>
<div class="highlight-python"><div class="highlight"><pre>    sequences per sample    iteration       PC.636  PC.635  PC.356  PC.481  PC.354  PC.593  PC.355  PC.607  PC.634
alpha_rarefaction_10_0.txt  10      0       7.0     10.0    6.0     8.0     9.0     9.0     7.0     9.0     10.0
alpha_rarefaction_10_1.txt  10      1       8.0     9.0     10.0    8.0     9.0     6.0     8.0     10.0    9.0
alpha_rarefaction_10_2.txt  10      2       8.0     10.0    9.0     10.0    8.0     8.0     9.0     10.0    7.0
alpha_rarefaction_10_3.txt  10      3       8.0     9.0     7.0     10.0    6.0     9.0     7.0     10.0    7.0
alpha_rarefaction_10_4.txt  10      4       8.0     9.0     8.0     7.0     8.0     10.0    9.0     9.0     9.0
</pre></div>
</div>
</div>
<div class="section" id="step-4-generate-alpha-rarefaction-plots">
<span id="generaterarecurves"></span><h3>Step 4. Generate alpha rarefaction plots<a class="headerlink" href="#step-4-generate-alpha-rarefaction-plots" title="Permalink to this headline">¶</a></h3>
<p>QIIME creates plots of alpha diversity vs. simulated sequencing effort, known as rarefaction plots, using the script <a class="reference external" href="../scripts/make_rarefaction_plots.html">make_rarefaction_plots.py</a>. This script takes a mapping file and any number of files generated by <a class="reference external" href="../scripts/collate_alpha.html">collate_alpha.py</a>, and creates alpha rarefaction curves. Each curve represents a sample and can be grouped by the sample metadata supplied in the mapping file.</p>
<p>This step generates a <tt class="file docutils literal"><span class="pre">arare/alpha_rarefaction_plots/rarefaction_plots.html</span></tt> file that can be opened with a web browser, in addition to other files. The <tt class="file docutils literal"><span class="pre">arare/alpha_rarefaction_plots/average_plots/</span></tt> folder contains the average plots for each metric and category.</p>
<p>The <tt class="file docutils literal"><span class="pre">arare/alpha_rarefaction_plots/average_tables/</span></tt> folder contains the diversity measure averages for each rarefied table, so the user can optionally plot the rarefaction curves in another application. To keep the results of <a class="reference external" href="../scripts/make_rarefaction_plots.html">make_rarefaction_plots.py</a> to a manageable size, these results are not generated unless you pass the <tt class="docutils literal"><span class="pre">--generate_average_tables</span></tt> option to <a class="reference external" href="../scripts/make_rarefaction_plots.html">make_rarefaction_plots.py</a>.</p>
</div>
<div class="section" id="viewing-alpha-rarefaction-plots">
<h3>Viewing alpha rarefaction plots<a class="headerlink" href="#viewing-alpha-rarefaction-plots" title="Permalink to this headline">¶</a></h3>
<p>To view the alpha rarefaction plots, open the file <tt class="file docutils literal"><span class="pre">arare/alpha_rarefaction_plots/rarefaction_plots.html</span></tt>. Once the browser window is open,  select the metric <cite>PD_whole_tree</cite> and the category <cite>Treatment</cite>, to reveal a plot like the figure below. You can click on the triangle next to each label in the legend to see all the samples that contribute to that category. Below each plot is a table displaying average values for each measure of alpha diversity for each group of samples in the specified category.</p>
<img alt="../_images/rarecurve.png" class="align-center" src="../_images/rarecurve.png" />
</div>
</div>
<div class="section" id="compute-beta-diversity-and-generate-ordination-plots">
<span id="compbetadivgenpcoa"></span><h2>Compute beta diversity and generate ordination plots<a class="headerlink" href="#compute-beta-diversity-and-generate-ordination-plots" title="Permalink to this headline">¶</a></h2>
<p>In addition to <em>alpha</em> (or <em>within-sample</em>) diversity, community ecologists are often interested in computing <em>beta</em> (or the <em>between-sample</em>) diversity between all pairs of samples in their study.</p>
<p>Beta diversity represents the explicit comparison of microbial (or other) communities based on their composition. Beta diversity metrics thus assess the differences between microbial communities. The fundamental output of these comparisons is a square, hollow matrix where a &#8220;distance&#8221; or dissimilarity is calculated between every pair of community samples, reflecting the dissimilarity between those samples. The data in this distance matrix can be visualized with analyses such as Principal Coordinates Analysis (PCoA) and hierarchical clustering.</p>
<p>Like alpha diversity, there are many possible beta diversity metrics that can be calculated with QIIME. The full list of metrics can be viewed by running:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">beta_diversity</span><span class="o">.</span><span class="n">py</span> <span class="o">-</span><span class="n">s</span>
</pre></div>
</div>
<p>Here, we will calculate beta diversity between our 9 microbial communities using the default beta diversity metrics of weighted and unweighted <a class="reference external" href="http://www.ncbi.nlm.nih.gov/pubmed/16332807">UniFrac</a>, which are phylogenetic measures used extensively in recent microbial community sequencing projects. To perform this analysis, we will use the <a class="reference external" href="../scripts/beta_diversity_through_plots.html">beta_diversity_through_plots.py</a> workflow, which performs the following steps:</p>
<ol class="arabic simple">
<li>Rarefy OTU table to remove sampling depth heterogeneity (<a class="reference external" href="../scripts/single_rarefaction.html">single_rarefaction.py</a>)</li>
<li>Compute beta diversity (<a class="reference external" href="../scripts/beta_diversity.html">beta_diversity.py</a>)</li>
<li>Run Principal Coordinates Analysis (<a class="reference external" href="../scripts/principal_coordinates.html">principal_coordinates.py</a>)</li>
<li>Generate Emperor PCoA plots (<a class="reference external" href="http://biocore.github.io/emperor/">make_emperor.py</a>)</li>
</ol>
<p>We can run the <a class="reference external" href="../scripts/beta_diversity_through_plots.html">beta_diversity_through_plots.py</a> workflow with the following command, which requires the OTU table (<tt class="docutils literal"><span class="pre">-i</span></tt>) and tree file (<tt class="docutils literal"><span class="pre">-t</span></tt>) from <a class="reference internal" href="#pickotusandrepseqs">above</a>, the metadata mapping file (<tt class="docutils literal"><span class="pre">-m</span></tt>), and the number of sequences per sample (<tt class="docutils literal"><span class="pre">-e</span></tt>, even sampling depth):</p>
<div class="highlight-python"><div class="highlight"><pre>beta_diversity_through_plots.py -i otus/otu_table.biom -m Fasting_Map.txt -o bdiv_even146 -t otus/rep_set.tre -e 146
</pre></div>
</div>
<p>Descriptions of the steps involved in <a class="reference external" href="../scripts/beta_diversity_through_plots.html">beta_diversity_through_plots.py</a> follow:</p>
<div class="section" id="step-1-rarefy-otu-table-to-remove-sampling-depth-heterogeneity">
<span id="compbetadiv"></span><h3>Step 1. Rarefy OTU table to remove sampling depth heterogeneity<a class="headerlink" href="#step-1-rarefy-otu-table-to-remove-sampling-depth-heterogeneity" title="Permalink to this headline">¶</a></h3>
<p>To remove sampling depth heterogeneity, we can perform rarefaction on our OTU table. Rarefaction is an ecological approach that allows users to standardize the data obtained from samples with different sequencing efforts, and to compare the OTU richness of the samples using these standardized data. For instance, if one of your samples yielded 10,000 sequences, and another yielded only 1,000 sequences, the species diversity within those samples may be much more influenced by sequencing effort than the underlying biology. We use rarefaction to randomly subsample the same number of sequences from each sample in order to compare the communities at a given level of sampling effort (an <em>even sampling depth</em>).</p>
<p>See the <tt class="docutils literal"><span class="pre">biom</span> <span class="pre">summarize-table</span></tt> section <a class="reference internal" href="#perlibrarystats">above</a> for the number of sequences in each of the 9 communities.</p>
<p>Since all samples have at least 146 sequences, a rarefaction level of 146 (specified by <tt class="docutils literal"><span class="pre">-e</span> <span class="pre">146</span></tt> above) allows us to compare all 9 samples at equal sequencing depth. Any samples containing fewer than 146 sequences would have been removed from these beta diversity analyses.</p>
</div>
<div class="section" id="step-2-compute-beta-diversity">
<h3>Step 2. Compute beta diversity<a class="headerlink" href="#step-2-compute-beta-diversity" title="Permalink to this headline">¶</a></h3>
<p>Beta diversity metrics assess the differences between microbial communities. By default, QIIME calculates both weighted and unweighted <a class="reference external" href="http://www.ncbi.nlm.nih.gov/pubmed/16332807">UniFrac</a>, which are phylogenetically-aware measures of beta diversity.</p>
<p>The resulting distance matrices (<tt class="file docutils literal"><span class="pre">bdiv_even146/unweighted_unifrac_dm.txt</span></tt> and <tt class="file docutils literal"><span class="pre">bdiv_even146/weighted_unifrac_dm.txt</span></tt>) are the basis for further analyses and visualizations (e.g., Principal Coordinates Analysis and hierarchical clustering).</p>
</div>
<div class="section" id="step-3-run-principal-coordinates-analysis">
<h3>Step 3. Run Principal Coordinates Analysis<a class="headerlink" href="#step-3-run-principal-coordinates-analysis" title="Permalink to this headline">¶</a></h3>
<p>Principal Coordinates Analysis (PCoA) is a technique that helps to extract and visualize a few highly-informative components of variation from complex, multidimensional data. This is a transformation that maps the samples present in the distance matrix to a new set of orthogonal axes such that a maximum amount of variation is explained by the first principal coordinate, the second largest amount of variation is explained by the second principal coordinate, etc. The principal coordinates can be plotted in two or three dimensions to provide an intuitive visualization of differences between samples. For more information on PCoA and other ordination techniques, see <a class="reference external" href="http://ordination.okstate.edu/">here</a>.</p>
<p>The files <tt class="file docutils literal"><span class="pre">bdiv_even146/unweighted_unifrac_pc.txt</span></tt> and <tt class="file docutils literal"><span class="pre">bdiv_even146/weighted_unifrac_pc.txt</span></tt> contain the results of PCoA applied to the unweighed and weighted <a class="reference external" href="http://www.ncbi.nlm.nih.gov/pubmed/16332807">UniFrac</a> distance matrices, respectively.</p>
</div>
<div class="section" id="step-4-generate-emperor-pcoa-plots">
<h3>Step 4. Generate Emperor PCoA plots<a class="headerlink" href="#step-4-generate-emperor-pcoa-plots" title="Permalink to this headline">¶</a></h3>
<p><a class="reference external" href="http://biocore.github.io/emperor/">Emperor</a> allows for the visualization of PCoA plots in three dimensions. HTML files are created in <tt class="file docutils literal"><span class="pre">bdiv_even146/unweighted_unifrac_emperor_pcoa_plot</span></tt> and <tt class="file docutils literal"><span class="pre">bdiv_even146/weighted_unifrac_emperor_pcoa_plot</span></tt> directories. For the <tt class="docutils literal"><span class="pre">Treatment</span></tt> column in the mapping file, all samples with the same treatment (control or fasted) will get the same color. By default, the five control samples are colored red and the four fasted samples are colored blue (colors, and many other plot settings, can be customized in Emperor). This allows you to easily visualize &#8220;clustering&#8221; by metadata category. The 3-D visualization software allows you to rotate the axes to see the data from different perspectives. By default, Emperor will plot the first three principal coordinates. Other combinations can be viewed using the &#8220;Axes&#8221; option in Emperor. All 8 components can be viewed using the left bottom menu &#8220;Parallel&#8221; using a parallel coordinates plot.</p>
<a class="reference internal image-reference" href="../_images/pcoa1.png"><img alt="../_images/pcoa1.png" class="align-center" src="../_images/pcoa1.png" style="width: 900px;" /></a>
</div>
</div>
<div class="section" id="jackknifed-beta-diversity-and-hierarchical-clustering">
<span id="jackbd"></span><span id="gendisthist"></span><h2>Jackknifed beta diversity and hierarchical clustering<a class="headerlink" href="#jackknifed-beta-diversity-and-hierarchical-clustering" title="Permalink to this headline">¶</a></h2>
<p>The <a class="reference external" href="../scripts/jackknifed_beta_diversity.html">jackknifed_beta_diversity.py</a> workflow uses jackknife replicates to estimate the uncertainty in PCoA plots and hierarchical clustering of microbial communities. Many of the same concepts relevant to beta diversity and PCoA are used here. <a class="reference external" href="../scripts/jackknifed_beta_diversity.html">jackknifed_beta_diversity.py</a> performs the following steps:</p>
<ol class="arabic simple">
<li>Compute beta diversity distance matrix from full OTU table and tree (<a class="reference external" href="../scripts/beta_diversity.html">beta_diversity.py</a>). These are only used if the user passed <tt class="docutils literal"><span class="pre">--master_tree</span> <span class="pre">full</span></tt>.</li>
<li>Build UPGMA tree from full distance matrix (<a class="reference external" href="../scripts/upgma_cluster.html">upgma_cluster.py</a>). This is only used if the user passed <tt class="docutils literal"><span class="pre">--master_tree</span> <span class="pre">full</span></tt>.</li>
<li>Build rarefied OTU tables (<a class="reference external" href="../scripts/multiple_rarefactions_even_depth.html">multiple_rarefactions_even_depth.py</a>).</li>
<li>Compute distance matrices from rarefied OTU tables (<a class="reference external" href="../scripts/beta_diversity.html">beta_diversity.py</a>).</li>
<li>Build UPGMA trees from rarefied distance matrices (<a class="reference external" href="../scripts/upgma_cluster.html">upgma_cluster.py</a>).</li>
<li>Compare rarefied UPGMA trees and determine jackknife support for tree nodes (<a class="reference external" href="../scripts/tree_compare.html">tree_compare.py</a> and <a class="reference external" href="../scripts/consensus_tree.html">consensus_tree.py</a>).</li>
<li>Compute PCoA on each rarefied distance matrix (<a class="reference external" href="../scripts/principal_coordinates.html">principal_coordinates.py</a>).</li>
<li>Compare rarefied PCoA plots from each rarefied distance matrix (<a class="reference external" href="http://biocore.github.io/emperor/">make_emperor.py</a>).</li>
</ol>
<p>We can run the workflow with the following command:</p>
<div class="highlight-python"><div class="highlight"><pre>jackknifed_beta_diversity.py -i otus/otu_table.biom -t otus/rep_set.tre -m Fasting_Map.txt -o jack -e 110
</pre></div>
</div>
<p>Descriptions of the steps involved in <a class="reference external" href="../scripts/jackknifed_beta_diversity.html">jackknifed_beta_diversity.py</a> follow:</p>
<div class="section" id="steps-1-and-2-upgma-clustering">
<span id="hiarchclust"></span><h3>Steps 1 and 2. UPGMA clustering<a class="headerlink" href="#steps-1-and-2-upgma-clustering" title="Permalink to this headline">¶</a></h3>
<p>Unweighted Pair Group Method with Arithmetic mean (UPGMA) is a type of hierarchical clustering method that uses average linkage. It can be used to interpret the distance matrix produced by <a class="reference external" href="../scripts/beta_diversity.html">beta_diversity.py</a>.</p>
<p>The output can be opened with tree viewing software, such as FigTree:</p>
<a class="reference internal image-reference" href="../_images/UPGMAbytreatment.png"><img alt="../_images/UPGMAbytreatment.png" class="align-center" src="../_images/UPGMAbytreatment.png" style="width: 700px;" /></a>
<p>This tree shows the relationship among the 9 samples, and reveals that the 4 samples from the guts of fasted mice cluster together (PC.6xx, fasting data is in <tt class="file docutils literal"><span class="pre">Fasting_Map.txt</span></tt>). The UPGMA trees resulting from this step should not be interpreted directly. Rather, they are only used in step 6 if the user passed <tt class="docutils literal"><span class="pre">--master_tree</span> <span class="pre">full</span></tt>. The default (<tt class="docutils literal"><span class="pre">--master_tree</span> <span class="pre">consensus</span></tt>) will result in a more biologically meaningful result, so these are provided only as a legacy option.</p>
</div>
<div class="section" id="steps-3-4-and-5-perform-jackknifing-support">
<span id="jacksupport"></span><h3>Steps 3, 4, and 5. Perform jackknifing support<a class="headerlink" href="#steps-3-4-and-5-perform-jackknifing-support" title="Permalink to this headline">¶</a></h3>
<p>To measure the robustness of our results to sequencing depth, we perform a jackknifing analysis, wherein a smaller number of sequences are chosen at random from each sample, and the resulting UPGMA tree from this subset of data is compared with the tree representing the entire data set. This process is repeated with many random subsets of data, and the tree nodes which prove more consistent across jackknifed datasets are deemed more robust.</p>
<p>First, the jackknifed OTU tables must be generated by subsampling the full data set. In this tutorial, each sample contains between 146 and 150 sequences, as shown with <a class="reference internal" href="#perlibrarystats">biom summarize-table</a>.</p>
<p>To ensure that a random subset of sequences is selected from each sample, we chose to select 110 sequences from each sample (75% of the smallest sample, though this value is only a guideline), which is designated by the <tt class="docutils literal"><span class="pre">-e</span></tt> option when running the workflow (see above).</p>
<p>More jackknife replicates provide a better estimate of the variability expected in beta diversity results, but at the cost of longer computational time. By default, QIIME generates 10 jackknife replicates of the available data. Each replicate is a simulation of a smaller sequencing effort (110 sequences in each sample, as defined above).</p>
<p>The workflow then computes a distance matrix for each jackknifed dataset, which results in 10 distance matrix files written to the <tt class="file docutils literal"><span class="pre">jack/unweighted_unifrac/rare_dm/</span></tt> directory and 10 distance matrices written to the <tt class="file docutils literal"><span class="pre">jack/weighted_unifrac/rare_dm/</span></tt> directory. Each set of 10 distance matrices is then used as input to hierarchical clustering with UPGMA, with the output written to the <tt class="file docutils literal"><span class="pre">jack/unweighted_unifrac/rare_upgma/</span></tt> and <tt class="file docutils literal"><span class="pre">jack/weighted_unifrac/rare_upgma/</span></tt> directories.</p>
</div>
<div class="section" id="step-6-compare-rarefied-upgma-trees-and-determine-jackknife-support-for-tree-nodes">
<span id="compjackclustertree"></span><h3>Step 6. Compare rarefied UPGMA trees and determine jackknife support for tree nodes<a class="headerlink" href="#step-6-compare-rarefied-upgma-trees-and-determine-jackknife-support-for-tree-nodes" title="Permalink to this headline">¶</a></h3>
<p>UPGMA clustering of the 10 distance matrices results in 10 hierarchical clusterings (or UPGMA trees) of the 9 mouse microbial communities, with each UPGMA tree being based on a random subsample of the available sequence data.</p>
<p>This step of the workflow compares each UPGMA tree to a consensus of those trees. Three files are written to <tt class="file docutils literal"><span class="pre">jack/unweighted_unifrac/upgma_cmp/</span></tt> and <tt class="file docutils literal"><span class="pre">jack/weighted_unifrac/upgma_cmp/</span></tt>:</p>
<blockquote>
<div><ul class="simple">
<li><tt class="file docutils literal"><span class="pre">master_tree.tre</span></tt>, which is identical to <tt class="file docutils literal"><span class="pre">jackknife_named_nodes.tre</span></tt> but each internal node of the UPGMA clustering is assigned a unique name</li>
<li><tt class="file docutils literal"><span class="pre">jackknife_named_nodes.tre</span></tt></li>
<li><tt class="file docutils literal"><span class="pre">jackknife_support.txt</span></tt> explains how frequently a given internal node had the same set of descendant samples in the jackknifed UPGMA clusters as it does in the UPGMA cluster using the full data set. A value of 0.5 indicates that half of the jackknifed data sets support that node, while 1.0 indicates perfect support.</li>
</ul>
</div></blockquote>
</div>
<div class="section" id="steps-7-and-8-compare-pcoa-plots">
<span id="comppcoa"></span><h3>Steps 7 and 8. Compare PCoA plots<a class="headerlink" href="#steps-7-and-8-compare-pcoa-plots" title="Permalink to this headline">¶</a></h3>
<p>The jackknifed replicate PCoA plots can be compared to assess the degree of variation from one replicate to the next. Emperor displays this variation by displaying confidence ellipsoids around the samples represented in a PCoA plot. The resulting plots are in <tt class="file docutils literal"><span class="pre">jack/unweighted_unifrac/emperor_pcoa_plots</span></tt> and <tt class="file docutils literal"><span class="pre">jack/weighted_unifrac/emperor_pcoa_plots</span></tt>:</p>
<a class="reference internal image-reference" href="../_images/jackpcoa.png"><img alt="../_images/jackpcoa.png" class="align-center" src="../_images/jackpcoa.png" style="width: 900px;" /></a>
</div>
<div class="section" id="generate-bootstrapped-tree">
<span id="genboottree"></span><h3>Generate bootstrapped tree<a class="headerlink" href="#generate-bootstrapped-tree" title="Permalink to this headline">¶</a></h3>
<p><tt class="file docutils literal"><span class="pre">jackknife_named_nodes.tre</span></tt> can be viewed with FigTree or another tree-viewing program. However, as an example, we can visualize the bootstrapped tree using QIIME&#8217;s <a class="reference external" href="../scripts/make_bootstrapped_tree.html">make_bootstrapped_tree.py</a>, as follows:</p>
<div class="highlight-python"><div class="highlight"><pre>make_bootstrapped_tree.py -m jack/unweighted_unifrac/upgma_cmp/master_tree.tre -s jack/unweighted_unifrac/upgma_cmp/jackknife_support.txt -o jack/unweighted_unifrac/upgma_cmp/jackknife_named_nodes.pdf
</pre></div>
</div>
<p>The resulting PDF shows the tree with internal nodes colored, red for 75-100% support, yellow for 50-75%, green for 25-50%, and blue for &lt; 25% support. Although UPGMA shows that PC.354 and PC.593 cluster together and PC.481 and PC.6xx cluster together, we cannot have high confidence in this result. However, there is excellent jackknife support for all fasted samples (PC.6xx) clustering separate from the non-fasted samples.</p>
<img alt="../_images/boottree.png" class="align-center" src="../_images/boottree.png" />
</div>
<div class="section" id="generate-3-d-biplots">
<h3>Generate 3-D biplots<a class="headerlink" href="#generate-3-d-biplots" title="Permalink to this headline">¶</a></h3>
<p>We can add taxa from the taxonomy tables in the <tt class="file docutils literal"><span class="pre">taxa_summary/</span></tt> directory to a 3-D PCoA plot using Emperor&#8217;s <a class="reference external" href="http://biocore.github.io/emperor/">make_emperor.py</a>. The coordinates of a given taxon are plotted as a weighted average of the coordinates of all samples, where the weights are the relative abundances of the taxon in the samples. The size of the sphere representing a taxon is proportional to the mean relative abundance of the taxon across all samples. The following command creates a biplot displaying the 5 most abundant class-level taxa:</p>
<div class="highlight-python"><div class="highlight"><pre>make_emperor.py -i bdiv_even146/unweighted_unifrac_pc.txt -m Fasting_Map.txt -t taxa_summary/otu_table_L3.txt --n_taxa_to_keep 5 -o biplots
</pre></div>
</div>
<p>The resulting html file <tt class="file docutils literal"><span class="pre">biplots/index.html</span></tt> shows a biplot similar to this:</p>
<img alt="../_images/biplot.png" class="align-center" src="../_images/biplot.png" />
</div>
</div>
<div class="section" id="running-workflow-scripts-in-parallel">
<h2>Running Workflow Scripts in Parallel<a class="headerlink" href="#running-workflow-scripts-in-parallel" title="Permalink to this headline">¶</a></h2>
<p>Most of QIIME&#8217;s workflows can be run in parallel. For information on how to use parallel QIIME, see <a class="reference external" href="parallel_qiime.html">here</a>.</p>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">

    
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">454 Overview Tutorial: de novo OTU picking and diversity analyses using 454 data</a><ul>
<li><a class="reference internal" href="#obtaining-the-data">Obtaining the data</a><ul>
<li><a class="reference internal" href="#sequences-fna">Sequences (.fna)</a></li>
<li><a class="reference internal" href="#quality-scores-qual">Quality Scores (.qual)</a></li>
<li><a class="reference internal" href="#mapping-file-tab-delimited-txt">Mapping File (Tab-delimited .txt)</a></li>
</ul>
</li>
<li><a class="reference internal" href="#validate-the-mapping-file">Validate the mapping file</a></li>
<li><a class="reference internal" href="#demultiplex-and-quality-filter-reads">Demultiplex and quality filter reads</a></li>
<li><a class="reference internal" href="#de-novo-otu-picking">De novo OTU picking</a><ul>
<li><a class="reference internal" href="#step-1-pick-otus-based-on-sequence-similarity-within-the-reads">Step 1. Pick OTUs based on sequence similarity within the reads</a></li>
<li><a class="reference internal" href="#step-2-pick-a-representative-sequence-for-each-otu">Step 2. Pick a representative sequence for each OTU</a></li>
<li><a class="reference internal" href="#step-3-assign-taxonomy-to-otu-representative-sequences">Step 3. Assign taxonomy to OTU representative sequences</a></li>
<li><a class="reference internal" href="#step-4-align-otu-representative-sequences">Step 4. Align OTU representative sequences</a></li>
<li><a class="reference internal" href="#step-5-filter-the-alignment">Step 5. Filter the alignment</a></li>
<li><a class="reference internal" href="#step-6-build-a-phylogenetic-tree">Step 6. Build a phylogenetic tree</a></li>
<li><a class="reference internal" href="#step-7-make-the-otu-table">Step 7. Make the OTU table</a></li>
</ul>
</li>
<li><a class="reference internal" href="#summarize-the-otu-table">Summarize the OTU table</a></li>
<li><a class="reference internal" href="#make-an-otu-network">Make an OTU network</a></li>
<li><a class="reference internal" href="#summarize-communities-by-taxonomic-composition">Summarize communities by taxonomic composition</a></li>
<li><a class="reference internal" href="#make-a-taxonomy-heatmap">Make a taxonomy heatmap</a></li>
<li><a class="reference internal" href="#compute-alpha-diversity-and-generate-alpha-rarefaction-plots">Compute alpha diversity and generate alpha rarefaction plots</a><ul>
<li><a class="reference internal" href="#step-1-generate-rarefied-otu-tables">Step 1. Generate rarefied OTU tables</a></li>
<li><a class="reference internal" href="#step-2-compute-measures-of-alpha-diversity-for-each-rarefied-otu-table">Step 2. Compute measures of alpha diversity for each rarefied OTU table</a></li>
<li><a class="reference internal" href="#step-3-collate-alpha-diversity-results">Step 3. Collate alpha diversity results</a></li>
<li><a class="reference internal" href="#step-4-generate-alpha-rarefaction-plots">Step 4. Generate alpha rarefaction plots</a></li>
<li><a class="reference internal" href="#viewing-alpha-rarefaction-plots">Viewing alpha rarefaction plots</a></li>
</ul>
</li>
<li><a class="reference internal" href="#compute-beta-diversity-and-generate-ordination-plots">Compute beta diversity and generate ordination plots</a><ul>
<li><a class="reference internal" href="#step-1-rarefy-otu-table-to-remove-sampling-depth-heterogeneity">Step 1. Rarefy OTU table to remove sampling depth heterogeneity</a></li>
<li><a class="reference internal" href="#step-2-compute-beta-diversity">Step 2. Compute beta diversity</a></li>
<li><a class="reference internal" href="#step-3-run-principal-coordinates-analysis">Step 3. Run Principal Coordinates Analysis</a></li>
<li><a class="reference internal" href="#step-4-generate-emperor-pcoa-plots">Step 4. Generate Emperor PCoA plots</a></li>
</ul>
</li>
<li><a class="reference internal" href="#jackknifed-beta-diversity-and-hierarchical-clustering">Jackknifed beta diversity and hierarchical clustering</a><ul>
<li><a class="reference internal" href="#steps-1-and-2-upgma-clustering">Steps 1 and 2. UPGMA clustering</a></li>
<li><a class="reference internal" href="#steps-3-4-and-5-perform-jackknifing-support">Steps 3, 4, and 5. Perform jackknifing support</a></li>
<li><a class="reference internal" href="#step-6-compare-rarefied-upgma-trees-and-determine-jackknife-support-for-tree-nodes">Step 6. Compare rarefied UPGMA trees and determine jackknife support for tree nodes</a></li>
<li><a class="reference internal" href="#steps-7-and-8-compare-pcoa-plots">Steps 7 and 8. Compare PCoA plots</a></li>
<li><a class="reference internal" href="#generate-bootstrapped-tree">Generate bootstrapped tree</a></li>
<li><a class="reference internal" href="#generate-3-d-biplots">Generate 3-D biplots</a></li>
</ul>
</li>
<li><a class="reference internal" href="#running-workflow-scripts-in-parallel">Running Workflow Scripts in Parallel</a></li>
</ul>
</li>
</ul>


    <h3><a href="../index.html">Site index</a></h3>
    <ul><li><ul>
    <li><a href="http://www.qiime.org">Home</a><br /></li>
    <li><a href="../install/index.html">Install</a><br /></li>
    <li><a href="index.html">Tutorials</a><br /></li>
    <li><a href="../scripts/index.html">Scripts</a><br /></li>
    <li><a href="http://help.qiime.org">Help</a><br /></li>
    <li><a href="http://qiime.org/home_static/dataFiles.html">Resources</a><br /></li>
    <li><a href="../documentation/index.html">File Formats</a><br /></li>
    <li><a href="http://workshops.qiime.org">Workshops</a><br /></li>
    <li><a href="http://qiime.wordpress.com">Blog</a><br /></li>
    <li><a href="../developer/index.html">Developer</a><br /></li>
    </ul></li></ul>


  <h4>Previous topic</h4>
  <p class="topless"><a href="illumina_overview_tutorial.html"
                        title="previous chapter">Illumina Overview Tutorial (an IPython Notebook): open reference OTU picking and core diversity analyses</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="open_reference_illumina_processing.html"
                        title="next chapter">Discussion of subsampled open-reference OTU picking in QIIME</a></p>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="open_reference_illumina_processing.html" title="Discussion of subsampled open-reference OTU picking in QIIME"
             >next</a> |</li>
        <li class="right" >
          <a href="illumina_overview_tutorial.html" title="Illumina Overview Tutorial (an IPython Notebook): open reference OTU picking and core diversity analyses"
             >previous</a> |</li>
        <li><a href="../index.html">Home</a> &raquo;</li>
          <li><a href="index.html" >QIIME Tutorials</a> &raquo;</li> 
      </ul>
    </div>
<div style="text-align: center; padding: 0px;"><br />
<a href="http://www.scikit-bio.org" target="_blank"><img src="../_static/powered_by_skbio.png" border="0" alt="sampledoc" height="70px"/></a>
<br /></div>

    <div class="footer">
        &copy; Copyright 2010-2014, QIIME Team.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.2.3.
    </div>
<script type="text/javascript">
var gaJsHost = (("https:" == document.location.protocol) ? "https://ssl." : "http://www.");
document.write(unescape("%3Cscript src='" + gaJsHost + "google-analytics.com/ga.js' type='text/javascript'%3E%3C/script%3E"));
</script>
<script type="text/javascript">
try {
var pageTracker = _gat._getTracker("UA-6636235-4");
pageTracker._trackPageview();
} catch(err) {}</script>

  </body>
</html>